// wdio-obsidian-service/dist/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __esm = (fn, res) => function __init() {
  return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])(fn = 0)), res;
};
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));

// ../../node_modules/.pnpm/tsup@8.5.1_tsx@4.21.0_typescript@5.9.3_yaml@2.8.2/node_modules/tsup/assets/esm_shims.js
import path from "path";
import { fileURLToPath } from "url";
var init_esm_shims = __esm({
  "../../node_modules/.pnpm/tsup@8.5.1_tsx@4.21.0_typescript@5.9.3_yaml@2.8.2/node_modules/tsup/assets/esm_shims.js"() {
    "use strict";
  }
});

// ../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/types/tokens.js
var require_tokens = __commonJS({
  "../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/types/tokens.js"(exports) {
    "use strict";
    init_esm_shims();
    Object.defineProperty(exports, "__esModule", { value: true });
  }
});

// ../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/types/types.js
var require_types = __commonJS({
  "../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/types/types.js"(exports) {
    "use strict";
    init_esm_shims();
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.types = void 0;
    var types;
    (function(types2) {
      types2[types2["ROOT"] = 0] = "ROOT";
      types2[types2["GROUP"] = 1] = "GROUP";
      types2[types2["POSITION"] = 2] = "POSITION";
      types2[types2["SET"] = 3] = "SET";
      types2[types2["RANGE"] = 4] = "RANGE";
      types2[types2["REPETITION"] = 5] = "REPETITION";
      types2[types2["REFERENCE"] = 6] = "REFERENCE";
      types2[types2["CHAR"] = 7] = "CHAR";
    })(types = exports.types || (exports.types = {}));
  }
});

// ../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/types/set-lookup.js
var require_set_lookup = __commonJS({
  "../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/types/set-lookup.js"(exports) {
    "use strict";
    init_esm_shims();
    Object.defineProperty(exports, "__esModule", { value: true });
  }
});

// ../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/types/index.js
var require_types2 = __commonJS({
  "../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/types/index.js"(exports) {
    "use strict";
    init_esm_shims();
    var __createBinding = exports && exports.__createBinding || (Object.create ? (function(o, m, k, k2) {
      if (k2 === void 0) k2 = k;
      Object.defineProperty(o, k2, { enumerable: true, get: function() {
        return m[k];
      } });
    }) : (function(o, m, k, k2) {
      if (k2 === void 0) k2 = k;
      o[k2] = m[k];
    }));
    var __exportStar = exports && exports.__exportStar || function(m, exports2) {
      for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports2, p)) __createBinding(exports2, m, p);
    };
    Object.defineProperty(exports, "__esModule", { value: true });
    __exportStar(require_tokens(), exports);
    __exportStar(require_types(), exports);
    __exportStar(require_set_lookup(), exports);
  }
});

// ../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/sets.js
var require_sets = __commonJS({
  "../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/sets.js"(exports) {
    "use strict";
    init_esm_shims();
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.anyChar = exports.notWhitespace = exports.whitespace = exports.notInts = exports.ints = exports.notWords = exports.words = void 0;
    var types_1 = require_types2();
    var INTS = () => [{ type: types_1.types.RANGE, from: 48, to: 57 }];
    var WORDS = () => [
      { type: types_1.types.CHAR, value: 95 },
      { type: types_1.types.RANGE, from: 97, to: 122 },
      { type: types_1.types.RANGE, from: 65, to: 90 },
      { type: types_1.types.RANGE, from: 48, to: 57 }
    ];
    var WHITESPACE = () => [
      { type: types_1.types.CHAR, value: 9 },
      { type: types_1.types.CHAR, value: 10 },
      { type: types_1.types.CHAR, value: 11 },
      { type: types_1.types.CHAR, value: 12 },
      { type: types_1.types.CHAR, value: 13 },
      { type: types_1.types.CHAR, value: 32 },
      { type: types_1.types.CHAR, value: 160 },
      { type: types_1.types.CHAR, value: 5760 },
      { type: types_1.types.RANGE, from: 8192, to: 8202 },
      { type: types_1.types.CHAR, value: 8232 },
      { type: types_1.types.CHAR, value: 8233 },
      { type: types_1.types.CHAR, value: 8239 },
      { type: types_1.types.CHAR, value: 8287 },
      { type: types_1.types.CHAR, value: 12288 },
      { type: types_1.types.CHAR, value: 65279 }
    ];
    var NOTANYCHAR = () => [
      { type: types_1.types.CHAR, value: 10 },
      { type: types_1.types.CHAR, value: 13 },
      { type: types_1.types.CHAR, value: 8232 },
      { type: types_1.types.CHAR, value: 8233 }
    ];
    exports.words = () => ({ type: types_1.types.SET, set: WORDS(), not: false });
    exports.notWords = () => ({ type: types_1.types.SET, set: WORDS(), not: true });
    exports.ints = () => ({ type: types_1.types.SET, set: INTS(), not: false });
    exports.notInts = () => ({ type: types_1.types.SET, set: INTS(), not: true });
    exports.whitespace = () => ({ type: types_1.types.SET, set: WHITESPACE(), not: false });
    exports.notWhitespace = () => ({ type: types_1.types.SET, set: WHITESPACE(), not: true });
    exports.anyChar = () => ({ type: types_1.types.SET, set: NOTANYCHAR(), not: true });
  }
});

// ../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/util.js
var require_util = __commonJS({
  "../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/util.js"(exports) {
    "use strict";
    init_esm_shims();
    var __createBinding = exports && exports.__createBinding || (Object.create ? (function(o, m, k, k2) {
      if (k2 === void 0) k2 = k;
      Object.defineProperty(o, k2, { enumerable: true, get: function() {
        return m[k];
      } });
    }) : (function(o, m, k, k2) {
      if (k2 === void 0) k2 = k;
      o[k2] = m[k];
    }));
    var __setModuleDefault = exports && exports.__setModuleDefault || (Object.create ? (function(o, v) {
      Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
      o["default"] = v;
    });
    var __importStar = exports && exports.__importStar || function(mod) {
      if (mod && mod.__esModule) return mod;
      var result = {};
      if (mod != null) {
        for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
      }
      __setModuleDefault(result, mod);
      return result;
    };
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.tokenizeClass = exports.strToChars = void 0;
    var types_1 = require_types2();
    var sets = __importStar(require_sets());
    var CTRL = "@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^ ?";
    exports.strToChars = (str) => {
      const charsRegex = /(\[\\b\])|(\\)?\\(?:u([A-F0-9]{4})|x([A-F0-9]{2})|c([@A-Z[\\\]^?])|([0tnvfr]))/g;
      return str.replace(charsRegex, (s, b, lbs, a16, b16, dctrl, eslsh) => {
        if (lbs) {
          return s;
        }
        let code = b ? 8 : a16 ? parseInt(a16, 16) : b16 ? parseInt(b16, 16) : dctrl ? CTRL.indexOf(dctrl) : {
          0: 0,
          t: 9,
          n: 10,
          v: 11,
          f: 12,
          r: 13
        }[eslsh];
        let c = String.fromCharCode(code);
        return /[[\]{}^$.|?*+()]/.test(c) ? `\\${c}` : c;
      });
    };
    exports.tokenizeClass = (str, regexpStr) => {
      var _a, _b, _c, _d, _e, _f, _g;
      let tokens = [], rs, c;
      const regexp = /\\(?:(w)|(d)|(s)|(W)|(D)|(S))|((?:(?:\\)(.)|([^\]\\]))-(((?:\\)])|(((?:\\)?([^\]])))))|(\])|(?:\\)?([^])/g;
      while ((rs = regexp.exec(str)) !== null) {
        const p = (_g = (_f = (_e = (_d = (_c = (_b = (_a = rs[1] && sets.words()) !== null && _a !== void 0 ? _a : rs[2] && sets.ints()) !== null && _b !== void 0 ? _b : rs[3] && sets.whitespace()) !== null && _c !== void 0 ? _c : rs[4] && sets.notWords()) !== null && _d !== void 0 ? _d : rs[5] && sets.notInts()) !== null && _e !== void 0 ? _e : rs[6] && sets.notWhitespace()) !== null && _f !== void 0 ? _f : rs[7] && {
          type: types_1.types.RANGE,
          from: (rs[8] || rs[9]).charCodeAt(0),
          to: (c = rs[10]).charCodeAt(c.length - 1)
        }) !== null && _g !== void 0 ? _g : (c = rs[16]) && { type: types_1.types.CHAR, value: c.charCodeAt(0) };
        if (p) {
          tokens.push(p);
        } else {
          return [tokens, regexp.lastIndex];
        }
      }
      throw new SyntaxError(`Invalid regular expression: /${regexpStr}/: Unterminated character class`);
    };
  }
});

// ../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/tokenizer.js
var require_tokenizer = __commonJS({
  "../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/tokenizer.js"(exports) {
    "use strict";
    init_esm_shims();
    var __createBinding = exports && exports.__createBinding || (Object.create ? (function(o, m, k, k2) {
      if (k2 === void 0) k2 = k;
      Object.defineProperty(o, k2, { enumerable: true, get: function() {
        return m[k];
      } });
    }) : (function(o, m, k, k2) {
      if (k2 === void 0) k2 = k;
      o[k2] = m[k];
    }));
    var __setModuleDefault = exports && exports.__setModuleDefault || (Object.create ? (function(o, v) {
      Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
      o["default"] = v;
    });
    var __importStar = exports && exports.__importStar || function(mod) {
      if (mod && mod.__esModule) return mod;
      var result = {};
      if (mod != null) {
        for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
      }
      __setModuleDefault(result, mod);
      return result;
    };
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.tokenizer = void 0;
    var util2 = __importStar(require_util());
    var types_1 = require_types2();
    var sets = __importStar(require_sets());
    var captureGroupFirstChar = /^[a-zA-Z_$]$/i;
    var captureGroupChars = /^[a-zA-Z0-9_$]$/i;
    var digit = /\d/;
    exports.tokenizer = (regexpStr) => {
      let i = 0, c;
      let start = { type: types_1.types.ROOT, stack: [] };
      let lastGroup = start;
      let last = start.stack;
      let groupStack = [];
      let referenceQueue = [];
      let groupCount = 0;
      const repeatErr = (col) => {
        throw new SyntaxError(`Invalid regular expression: /${regexpStr}/: Nothing to repeat at column ${col - 1}`);
      };
      let str = util2.strToChars(regexpStr);
      while (i < str.length) {
        switch (c = str[i++]) {
          // Handle escaped characters, inclues a few sets.
          case "\\":
            if (i === str.length) {
              throw new SyntaxError(`Invalid regular expression: /${regexpStr}/: \\ at end of pattern`);
            }
            switch (c = str[i++]) {
              case "b":
                last.push({ type: types_1.types.POSITION, value: "b" });
                break;
              case "B":
                last.push({ type: types_1.types.POSITION, value: "B" });
                break;
              case "w":
                last.push(sets.words());
                break;
              case "W":
                last.push(sets.notWords());
                break;
              case "d":
                last.push(sets.ints());
                break;
              case "D":
                last.push(sets.notInts());
                break;
              case "s":
                last.push(sets.whitespace());
                break;
              case "S":
                last.push(sets.notWhitespace());
                break;
              default:
                if (digit.test(c)) {
                  let digits = c;
                  while (i < str.length && digit.test(str[i])) {
                    digits += str[i++];
                  }
                  let value = parseInt(digits, 10);
                  const reference = { type: types_1.types.REFERENCE, value };
                  last.push(reference);
                  referenceQueue.push({ reference, stack: last, index: last.length - 1 });
                } else {
                  last.push({ type: types_1.types.CHAR, value: c.charCodeAt(0) });
                }
            }
            break;
          // Positionals.
          case "^":
            last.push({ type: types_1.types.POSITION, value: "^" });
            break;
          case "$":
            last.push({ type: types_1.types.POSITION, value: "$" });
            break;
          // Handle custom sets.
          case "[": {
            let not;
            if (str[i] === "^") {
              not = true;
              i++;
            } else {
              not = false;
            }
            let classTokens = util2.tokenizeClass(str.slice(i), regexpStr);
            i += classTokens[1];
            last.push({
              type: types_1.types.SET,
              set: classTokens[0],
              not
            });
            break;
          }
          // Class of any character except \n.
          case ".":
            last.push(sets.anyChar());
            break;
          // Push group onto stack.
          case "(": {
            let group = {
              type: types_1.types.GROUP,
              stack: [],
              remember: true
            };
            if (str[i] === "?") {
              c = str[i + 1];
              i += 2;
              if (c === "=") {
                group.followedBy = true;
                group.remember = false;
              } else if (c === "!") {
                group.notFollowedBy = true;
                group.remember = false;
              } else if (c === "<") {
                let name = "";
                if (captureGroupFirstChar.test(str[i])) {
                  name += str[i];
                  i++;
                } else {
                  throw new SyntaxError(`Invalid regular expression: /${regexpStr}/: Invalid capture group name, character '${str[i]}' after '<' at column ${i + 1}`);
                }
                while (i < str.length && captureGroupChars.test(str[i])) {
                  name += str[i];
                  i++;
                }
                if (!name) {
                  throw new SyntaxError(`Invalid regular expression: /${regexpStr}/: Invalid capture group name, character '${str[i]}' after '<' at column ${i + 1}`);
                }
                if (str[i] !== ">") {
                  throw new SyntaxError(`Invalid regular expression: /${regexpStr}/: Unclosed capture group name, expected '>', found '${str[i]}' at column ${i + 1}`);
                }
                group.name = name;
                i++;
              } else if (c === ":") {
                group.remember = false;
              } else {
                throw new SyntaxError(`Invalid regular expression: /${regexpStr}/: Invalid group, character '${c}' after '?' at column ${i - 1}`);
              }
            } else {
              groupCount += 1;
            }
            last.push(group);
            groupStack.push(lastGroup);
            lastGroup = group;
            last = group.stack;
            break;
          }
          // Pop group out of stack.
          case ")":
            if (groupStack.length === 0) {
              throw new SyntaxError(`Invalid regular expression: /${regexpStr}/: Unmatched ) at column ${i - 1}`);
            }
            lastGroup = groupStack.pop();
            last = lastGroup.options ? lastGroup.options[lastGroup.options.length - 1] : lastGroup.stack;
            break;
          // Use pipe character to give more choices.
          case "|": {
            if (!lastGroup.options) {
              lastGroup.options = [lastGroup.stack];
              delete lastGroup.stack;
            }
            let stack = [];
            lastGroup.options.push(stack);
            last = stack;
            break;
          }
          // Repetition.
          // For every repetition, remove last element from last stack
          // then insert back a RANGE object.
          // This design is chosen because there could be more than
          // one repetition symbols in a regex i.e. `a?+{2,3}`.
          case "{": {
            let rs = /^(\d+)(,(\d+)?)?\}/.exec(str.slice(i)), min, max;
            if (rs !== null) {
              if (last.length === 0) {
                repeatErr(i);
              }
              min = parseInt(rs[1], 10);
              max = rs[2] ? rs[3] ? parseInt(rs[3], 10) : Infinity : min;
              i += rs[0].length;
              last.push({
                type: types_1.types.REPETITION,
                min,
                max,
                value: last.pop()
              });
            } else {
              last.push({
                type: types_1.types.CHAR,
                value: 123
              });
            }
            break;
          }
          case "?":
            if (last.length === 0) {
              repeatErr(i);
            }
            last.push({
              type: types_1.types.REPETITION,
              min: 0,
              max: 1,
              value: last.pop()
            });
            break;
          case "+":
            if (last.length === 0) {
              repeatErr(i);
            }
            last.push({
              type: types_1.types.REPETITION,
              min: 1,
              max: Infinity,
              value: last.pop()
            });
            break;
          case "*":
            if (last.length === 0) {
              repeatErr(i);
            }
            last.push({
              type: types_1.types.REPETITION,
              min: 0,
              max: Infinity,
              value: last.pop()
            });
            break;
          // Default is a character that is not `\[](){}?+*^$`.
          default:
            last.push({
              type: types_1.types.CHAR,
              value: c.charCodeAt(0)
            });
        }
      }
      if (groupStack.length !== 0) {
        throw new SyntaxError(`Invalid regular expression: /${regexpStr}/: Unterminated group`);
      }
      updateReferences(referenceQueue, groupCount);
      return start;
    };
    function updateReferences(referenceQueue, groupCount) {
      for (const elem of referenceQueue.reverse()) {
        if (groupCount < elem.reference.value) {
          elem.reference.type = types_1.types.CHAR;
          const valueString = elem.reference.value.toString();
          elem.reference.value = parseInt(valueString, 8);
          if (!/^[0-7]+$/.test(valueString)) {
            let i = 0;
            while (valueString[i] !== "8" && valueString[i] !== "9") {
              i += 1;
            }
            if (i === 0) {
              elem.reference.value = valueString.charCodeAt(0);
              i += 1;
            } else {
              elem.reference.value = parseInt(valueString.slice(0, i), 8);
            }
            if (valueString.length > i) {
              const tail = elem.stack.splice(elem.index + 1);
              for (const char of valueString.slice(i)) {
                elem.stack.push({
                  type: types_1.types.CHAR,
                  value: char.charCodeAt(0)
                });
              }
              elem.stack.push(...tail);
            }
          }
        }
      }
    }
  }
});

// ../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/sets-lookup.js
var require_sets_lookup = __commonJS({
  "../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/sets-lookup.js"(exports) {
    "use strict";
    init_esm_shims();
    var __createBinding = exports && exports.__createBinding || (Object.create ? (function(o, m, k, k2) {
      if (k2 === void 0) k2 = k;
      Object.defineProperty(o, k2, { enumerable: true, get: function() {
        return m[k];
      } });
    }) : (function(o, m, k, k2) {
      if (k2 === void 0) k2 = k;
      o[k2] = m[k];
    }));
    var __setModuleDefault = exports && exports.__setModuleDefault || (Object.create ? (function(o, v) {
      Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
      o["default"] = v;
    });
    var __importStar = exports && exports.__importStar || function(mod) {
      if (mod && mod.__esModule) return mod;
      var result = {};
      if (mod != null) {
        for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
      }
      __setModuleDefault(result, mod);
      return result;
    };
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.NOTANYCHAR = exports.WHITESPACE = exports.WORDS = exports.INTS = void 0;
    var Sets = __importStar(require_sets());
    var types_1 = require_types2();
    function setToLookup(tokens) {
      let lookup = {};
      let len = 0;
      for (const token of tokens) {
        if (token.type === types_1.types.CHAR) {
          lookup[token.value] = true;
        }
        if (token.type === types_1.types.RANGE) {
          lookup[`${token.from}-${token.to}`] = true;
        }
        len += 1;
      }
      return {
        lookup: () => Object.assign({}, lookup),
        len
      };
    }
    exports.INTS = setToLookup(Sets.ints().set);
    exports.WORDS = setToLookup(Sets.words().set);
    exports.WHITESPACE = setToLookup(Sets.whitespace().set);
    exports.NOTANYCHAR = setToLookup(Sets.anyChar().set);
  }
});

// ../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/write-set-tokens.js
var require_write_set_tokens = __commonJS({
  "../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/write-set-tokens.js"(exports) {
    "use strict";
    init_esm_shims();
    var __createBinding = exports && exports.__createBinding || (Object.create ? (function(o, m, k, k2) {
      if (k2 === void 0) k2 = k;
      Object.defineProperty(o, k2, { enumerable: true, get: function() {
        return m[k];
      } });
    }) : (function(o, m, k, k2) {
      if (k2 === void 0) k2 = k;
      o[k2] = m[k];
    }));
    var __setModuleDefault = exports && exports.__setModuleDefault || (Object.create ? (function(o, v) {
      Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
      o["default"] = v;
    });
    var __importStar = exports && exports.__importStar || function(mod) {
      if (mod && mod.__esModule) return mod;
      var result = {};
      if (mod != null) {
        for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
      }
      __setModuleDefault(result, mod);
      return result;
    };
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.writeSetTokens = exports.setChar = void 0;
    var types_1 = require_types2();
    var sets = __importStar(require_sets_lookup());
    function setChar(charCode) {
      return charCode === 94 ? "\\^" : charCode === 92 ? "\\\\" : charCode === 93 ? "\\]" : charCode === 45 ? "\\-" : String.fromCharCode(charCode);
    }
    exports.setChar = setChar;
    function isSameSet(set, { lookup, len }) {
      if (len !== set.length) {
        return false;
      }
      const map = lookup();
      for (const elem of set) {
        if (elem.type === types_1.types.SET) {
          return false;
        }
        const key = elem.type === types_1.types.CHAR ? elem.value : `${elem.from}-${elem.to}`;
        if (map[key]) {
          map[key] = false;
        } else {
          return false;
        }
      }
      return true;
    }
    function writeSetTokens(set, isNested = false) {
      if (isSameSet(set.set, sets.INTS)) {
        return set.not ? "\\D" : "\\d";
      }
      if (isSameSet(set.set, sets.WORDS)) {
        return set.not ? "\\W" : "\\w";
      }
      if (set.not && isSameSet(set.set, sets.NOTANYCHAR)) {
        return ".";
      }
      if (isSameSet(set.set, sets.WHITESPACE)) {
        return set.not ? "\\S" : "\\s";
      }
      let tokenString = "";
      for (let i = 0; i < set.set.length; i++) {
        const subset = set.set[i];
        tokenString += writeSetToken(subset);
      }
      const contents = `${set.not ? "^" : ""}${tokenString}`;
      return isNested ? contents : `[${contents}]`;
    }
    exports.writeSetTokens = writeSetTokens;
    function writeSetToken(set) {
      if (set.type === types_1.types.CHAR) {
        return setChar(set.value);
      } else if (set.type === types_1.types.RANGE) {
        return `${setChar(set.from)}-${setChar(set.to)}`;
      }
      return writeSetTokens(set, true);
    }
  }
});

// ../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/reconstruct.js
var require_reconstruct = __commonJS({
  "../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/reconstruct.js"(exports) {
    "use strict";
    init_esm_shims();
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.reconstruct = void 0;
    var types_1 = require_types2();
    var write_set_tokens_1 = require_write_set_tokens();
    var reduceStack = (stack) => stack.map(exports.reconstruct).join("");
    var createAlternate = (token) => {
      if ("options" in token) {
        return token.options.map(reduceStack).join("|");
      } else if ("stack" in token) {
        return reduceStack(token.stack);
      } else {
        throw new Error(`options or stack must be Root or Group token`);
      }
    };
    exports.reconstruct = (token) => {
      switch (token.type) {
        case types_1.types.ROOT:
          return createAlternate(token);
        case types_1.types.CHAR: {
          const c = String.fromCharCode(token.value);
          return (/[[\\{}$^.|?*+()]/.test(c) ? "\\" : "") + c;
        }
        case types_1.types.POSITION:
          if (token.value === "^" || token.value === "$") {
            return token.value;
          } else {
            return `\\${token.value}`;
          }
        case types_1.types.REFERENCE:
          return `\\${token.value}`;
        case types_1.types.SET:
          return write_set_tokens_1.writeSetTokens(token);
        case types_1.types.GROUP: {
          const prefix2 = token.name ? `?<${token.name}>` : token.remember ? "" : token.followedBy ? "?=" : token.notFollowedBy ? "?!" : "?:";
          return `(${prefix2}${createAlternate(token)})`;
        }
        case types_1.types.REPETITION: {
          const { min, max } = token;
          let endWith;
          if (min === 0 && max === 1) {
            endWith = "?";
          } else if (min === 1 && max === Infinity) {
            endWith = "+";
          } else if (min === 0 && max === Infinity) {
            endWith = "*";
          } else if (max === Infinity) {
            endWith = `{${min},}`;
          } else if (min === max) {
            endWith = `{${min}}`;
          } else {
            endWith = `{${min},${max}}`;
          }
          return `${exports.reconstruct(token.value)}${endWith}`;
        }
        case types_1.types.RANGE:
          return `${write_set_tokens_1.setChar(token.from)}-${write_set_tokens_1.setChar(token.to)}`;
        default:
          throw new Error(`Invalid token type ${token}`);
      }
    };
  }
});

// ../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/index.js
var require_dist = __commonJS({
  "../../node_modules/.pnpm/ret@0.5.0/node_modules/ret/dist/index.js"(exports, module) {
    "use strict";
    init_esm_shims();
    var __createBinding = exports && exports.__createBinding || (Object.create ? (function(o, m, k, k2) {
      if (k2 === void 0) k2 = k;
      Object.defineProperty(o, k2, { enumerable: true, get: function() {
        return m[k];
      } });
    }) : (function(o, m, k, k2) {
      if (k2 === void 0) k2 = k;
      o[k2] = m[k];
    }));
    var __exportStar = exports && exports.__exportStar || function(m, exports2) {
      for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports2, p)) __createBinding(exports2, m, p);
    };
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.types = void 0;
    var types_1 = require_types2();
    Object.defineProperty(exports, "types", { enumerable: true, get: function() {
      return types_1.types;
    } });
    __exportStar(require_tokenizer(), exports);
    __exportStar(require_reconstruct(), exports);
    var tokenizer_1 = require_tokenizer();
    var reconstruct_1 = require_reconstruct();
    __exportStar(require_types2(), exports);
    exports.default = tokenizer_1.tokenizer;
    module.exports = tokenizer_1.tokenizer;
    module.exports.types = types_1.types;
    module.exports.reconstruct = reconstruct_1.reconstruct;
  }
});

// ../../node_modules/.pnpm/safe-regex2@5.1.0/node_modules/safe-regex2/index.js
var require_safe_regex2 = __commonJS({
  "../../node_modules/.pnpm/safe-regex2@5.1.0/node_modules/safe-regex2/index.js"(exports, module) {
    "use strict";
    init_esm_shims();
    var parse = require_dist();
    var { types } = require_dist();
    function walk(node, opts, starHeight) {
      let i;
      let ok;
      let len;
      if (node.type === types.REPETITION) {
        starHeight++;
        opts.reps++;
        if (starHeight > 1) return false;
        if (opts.reps > opts.limit) return false;
      }
      if (node.options) {
        for (i = 0, len = node.options.length; i < len; i++) {
          ok = walk({ stack: node.options[i] }, opts, starHeight);
          if (!ok) return false;
        }
      }
      const stack = node.stack || node.value?.stack;
      if (!stack) return true;
      for (i = 0, len = stack.length; i < len; i++) {
        ok = walk(stack[i], opts, starHeight);
        if (!ok) return false;
      }
      return true;
    }
    function safeRegex(re, options) {
      const opts = {
        reps: 0,
        limit: options?.limit ?? 25
      };
      if (isRegExp(re)) re = re.source;
      else if (typeof re !== "string") re = String(re);
      try {
        return walk(parse(re), opts, 0);
      } catch {
        return false;
      }
    }
    function isRegExp(x) {
      return Object.prototype.toString.call(x) === "[object RegExp]";
    }
    module.exports = safeRegex;
    module.exports.default = safeRegex;
    module.exports.safeRegex = safeRegex;
  }
});

// ../../node_modules/.pnpm/loglevel@1.9.2/node_modules/loglevel/lib/loglevel.js
var require_loglevel = __commonJS({
  "../../node_modules/.pnpm/loglevel@1.9.2/node_modules/loglevel/lib/loglevel.js"(exports, module) {
    "use strict";
    init_esm_shims();
    (function(root, definition) {
      "use strict";
      if (typeof define === "function" && define.amd) {
        define(definition);
      } else if (typeof module === "object" && module.exports) {
        module.exports = definition();
      } else {
        root.log = definition();
      }
    })(exports, function() {
      "use strict";
      var noop = function() {
      };
      var undefinedType = "undefined";
      var isIE = typeof window !== undefinedType && typeof window.navigator !== undefinedType && /Trident\/|MSIE /.test(window.navigator.userAgent);
      var logMethods = [
        "trace",
        "debug",
        "info",
        "warn",
        "error"
      ];
      var _loggersByName = {};
      var defaultLogger = null;
      function bindMethod(obj, methodName) {
        var method = obj[methodName];
        if (typeof method.bind === "function") {
          return method.bind(obj);
        } else {
          try {
            return Function.prototype.bind.call(method, obj);
          } catch (e) {
            return function() {
              return Function.prototype.apply.apply(method, [obj, arguments]);
            };
          }
        }
      }
      function traceForIE() {
        if (console.log) {
          if (console.log.apply) {
            console.log.apply(console, arguments);
          } else {
            Function.prototype.apply.apply(console.log, [console, arguments]);
          }
        }
        if (console.trace) console.trace();
      }
      function realMethod(methodName) {
        if (methodName === "debug") {
          methodName = "log";
        }
        if (typeof console === undefinedType) {
          return false;
        } else if (methodName === "trace" && isIE) {
          return traceForIE;
        } else if (console[methodName] !== void 0) {
          return bindMethod(console, methodName);
        } else if (console.log !== void 0) {
          return bindMethod(console, "log");
        } else {
          return noop;
        }
      }
      function replaceLoggingMethods() {
        var level = this.getLevel();
        for (var i = 0; i < logMethods.length; i++) {
          var methodName = logMethods[i];
          this[methodName] = i < level ? noop : this.methodFactory(methodName, level, this.name);
        }
        this.log = this.debug;
        if (typeof console === undefinedType && level < this.levels.SILENT) {
          return "No console available for logging";
        }
      }
      function enableLoggingWhenConsoleArrives(methodName) {
        return function() {
          if (typeof console !== undefinedType) {
            replaceLoggingMethods.call(this);
            this[methodName].apply(this, arguments);
          }
        };
      }
      function defaultMethodFactory(methodName, _level, _loggerName) {
        return realMethod(methodName) || enableLoggingWhenConsoleArrives.apply(this, arguments);
      }
      function Logger(name, factory) {
        var self = this;
        var inheritedLevel;
        var defaultLevel;
        var userLevel;
        var storageKey = "loglevel";
        if (typeof name === "string") {
          storageKey += ":" + name;
        } else if (typeof name === "symbol") {
          storageKey = void 0;
        }
        function persistLevelIfPossible(levelNum) {
          var levelName = (logMethods[levelNum] || "silent").toUpperCase();
          if (typeof window === undefinedType || !storageKey) return;
          try {
            window.localStorage[storageKey] = levelName;
            return;
          } catch (ignore) {
          }
          try {
            window.document.cookie = encodeURIComponent(storageKey) + "=" + levelName + ";";
          } catch (ignore) {
          }
        }
        function getPersistedLevel() {
          var storedLevel;
          if (typeof window === undefinedType || !storageKey) return;
          try {
            storedLevel = window.localStorage[storageKey];
          } catch (ignore) {
          }
          if (typeof storedLevel === undefinedType) {
            try {
              var cookie = window.document.cookie;
              var cookieName = encodeURIComponent(storageKey);
              var location2 = cookie.indexOf(cookieName + "=");
              if (location2 !== -1) {
                storedLevel = /^([^;]+)/.exec(
                  cookie.slice(location2 + cookieName.length + 1)
                )[1];
              }
            } catch (ignore) {
            }
          }
          if (self.levels[storedLevel] === void 0) {
            storedLevel = void 0;
          }
          return storedLevel;
        }
        function clearPersistedLevel() {
          if (typeof window === undefinedType || !storageKey) return;
          try {
            window.localStorage.removeItem(storageKey);
          } catch (ignore) {
          }
          try {
            window.document.cookie = encodeURIComponent(storageKey) + "=; expires=Thu, 01 Jan 1970 00:00:00 UTC";
          } catch (ignore) {
          }
        }
        function normalizeLevel(input) {
          var level = input;
          if (typeof level === "string" && self.levels[level.toUpperCase()] !== void 0) {
            level = self.levels[level.toUpperCase()];
          }
          if (typeof level === "number" && level >= 0 && level <= self.levels.SILENT) {
            return level;
          } else {
            throw new TypeError("log.setLevel() called with invalid level: " + input);
          }
        }
        self.name = name;
        self.levels = {
          "TRACE": 0,
          "DEBUG": 1,
          "INFO": 2,
          "WARN": 3,
          "ERROR": 4,
          "SILENT": 5
        };
        self.methodFactory = factory || defaultMethodFactory;
        self.getLevel = function() {
          if (userLevel != null) {
            return userLevel;
          } else if (defaultLevel != null) {
            return defaultLevel;
          } else {
            return inheritedLevel;
          }
        };
        self.setLevel = function(level, persist) {
          userLevel = normalizeLevel(level);
          if (persist !== false) {
            persistLevelIfPossible(userLevel);
          }
          return replaceLoggingMethods.call(self);
        };
        self.setDefaultLevel = function(level) {
          defaultLevel = normalizeLevel(level);
          if (!getPersistedLevel()) {
            self.setLevel(level, false);
          }
        };
        self.resetLevel = function() {
          userLevel = null;
          clearPersistedLevel();
          replaceLoggingMethods.call(self);
        };
        self.enableAll = function(persist) {
          self.setLevel(self.levels.TRACE, persist);
        };
        self.disableAll = function(persist) {
          self.setLevel(self.levels.SILENT, persist);
        };
        self.rebuild = function() {
          if (defaultLogger !== self) {
            inheritedLevel = normalizeLevel(defaultLogger.getLevel());
          }
          replaceLoggingMethods.call(self);
          if (defaultLogger === self) {
            for (var childName in _loggersByName) {
              _loggersByName[childName].rebuild();
            }
          }
        };
        inheritedLevel = normalizeLevel(
          defaultLogger ? defaultLogger.getLevel() : "WARN"
        );
        var initialLevel = getPersistedLevel();
        if (initialLevel != null) {
          userLevel = normalizeLevel(initialLevel);
        }
        replaceLoggingMethods.call(self);
      }
      defaultLogger = new Logger();
      defaultLogger.getLogger = function getLogger2(name) {
        if (typeof name !== "symbol" && typeof name !== "string" || name === "") {
          throw new TypeError("You must supply a name when creating a logger.");
        }
        var logger = _loggersByName[name];
        if (!logger) {
          logger = _loggersByName[name] = new Logger(
            name,
            defaultLogger.methodFactory
          );
        }
        return logger;
      };
      var _log = typeof window !== undefinedType ? window.log : void 0;
      defaultLogger.noConflict = function() {
        if (typeof window !== undefinedType && window.log === defaultLogger) {
          window.log = _log;
        }
        return defaultLogger;
      };
      defaultLogger.getLoggers = function getLoggers() {
        return _loggersByName;
      };
      defaultLogger["default"] = defaultLogger;
      return defaultLogger;
    });
  }
});

// ../../node_modules/.pnpm/loglevel-plugin-prefix@0.8.4/node_modules/loglevel-plugin-prefix/lib/loglevel-plugin-prefix.js
var require_loglevel_plugin_prefix = __commonJS({
  "../../node_modules/.pnpm/loglevel-plugin-prefix@0.8.4/node_modules/loglevel-plugin-prefix/lib/loglevel-plugin-prefix.js"(exports, module) {
    "use strict";
    init_esm_shims();
    (function(root, factory) {
      if (typeof define === "function" && define.amd) {
        define(factory);
      } else if (typeof module === "object" && module.exports) {
        module.exports = factory();
      } else {
        root.prefix = factory(root);
      }
    })(exports, function(root) {
      "use strict";
      var merge = function(target) {
        var i = 1;
        var length = arguments.length;
        var key;
        for (; i < length; i++) {
          for (key in arguments[i]) {
            if (Object.prototype.hasOwnProperty.call(arguments[i], key)) {
              target[key] = arguments[i][key];
            }
          }
        }
        return target;
      };
      var defaults = {
        template: "[%t] %l:",
        levelFormatter: function(level) {
          return level.toUpperCase();
        },
        nameFormatter: function(name) {
          return name || "root";
        },
        timestampFormatter: function(date) {
          return date.toTimeString().replace(/.*(\d{2}:\d{2}:\d{2}).*/, "$1");
        },
        format: void 0
      };
      var loglevel;
      var configs = {};
      var reg = function(rootLogger) {
        if (!rootLogger || !rootLogger.getLogger) {
          throw new TypeError("Argument is not a root logger");
        }
        loglevel = rootLogger;
      };
      var apply = function(logger, config) {
        if (!logger || !logger.setLevel) {
          throw new TypeError("Argument is not a logger");
        }
        var originalFactory2 = logger.methodFactory;
        var name = logger.name || "";
        var parent2 = configs[name] || configs[""] || defaults;
        function methodFactory(methodName, logLevel, loggerName) {
          var originalMethod = originalFactory2(methodName, logLevel, loggerName);
          var options = configs[loggerName] || configs[""];
          var hasTimestamp = options.template.indexOf("%t") !== -1;
          var hasLevel = options.template.indexOf("%l") !== -1;
          var hasName = options.template.indexOf("%n") !== -1;
          return function() {
            var content = "";
            var length = arguments.length;
            var args = Array(length);
            var key = 0;
            for (; key < length; key++) {
              args[key] = arguments[key];
            }
            if (name || !configs[loggerName]) {
              var timestamp = options.timestampFormatter(/* @__PURE__ */ new Date());
              var level = options.levelFormatter(methodName);
              var lname = options.nameFormatter(loggerName);
              if (options.format) {
                content += options.format(level, lname, timestamp);
              } else {
                content += options.template;
                if (hasTimestamp) {
                  content = content.replace(/%t/, timestamp);
                }
                if (hasLevel) content = content.replace(/%l/, level);
                if (hasName) content = content.replace(/%n/, lname);
              }
              if (args.length && typeof args[0] === "string") {
                args[0] = content + " " + args[0];
              } else {
                args.unshift(content);
              }
            }
            originalMethod.apply(void 0, args);
          };
        }
        if (!configs[name]) {
          logger.methodFactory = methodFactory;
        }
        config = config || {};
        if (config.template) config.format = void 0;
        configs[name] = merge({}, parent2, config);
        logger.setLevel(logger.getLevel());
        if (!loglevel) {
          logger.warn(
            "It is necessary to call the function reg() of loglevel-plugin-prefix before calling apply. From the next release, it will throw an error. See more: https://github.com/kutuluk/loglevel-plugin-prefix/blob/master/README.md"
          );
        }
        return logger;
      };
      var api = {
        reg,
        apply
      };
      var save;
      if (root) {
        save = root.prefix;
        api.noConflict = function() {
          if (root.prefix === api) {
            root.prefix = save;
          }
          return api;
        };
      }
      return api;
    });
  }
});

// src/index.ts
init_esm_shims();
import ObsidianLauncher2 from "obsidian-launcher";
import { deprecate } from "util";

// src/service.ts
init_esm_shims();
import fs2 from "fs";
import fsAsync3 from "fs/promises";
import path4 from "path";
import { SevereServiceError } from "webdriverio";

// ../../node_modules/.pnpm/@wdio+logger@9.18.0/node_modules/@wdio/logger/build/index.js
init_esm_shims();
var import_safe_regex2 = __toESM(require_safe_regex2(), 1);
var import_loglevel = __toESM(require_loglevel(), 1);
import fs from "fs";
import util from "util";

// ../../node_modules/.pnpm/chalk@5.6.2/node_modules/chalk/source/index.js
init_esm_shims();

// ../../node_modules/.pnpm/chalk@5.6.2/node_modules/chalk/source/vendor/ansi-styles/index.js
init_esm_shims();
var ANSI_BACKGROUND_OFFSET = 10;
var wrapAnsi16 = (offset = 0) => (code) => `\x1B[${code + offset}m`;
var wrapAnsi256 = (offset = 0) => (code) => `\x1B[${38 + offset};5;${code}m`;
var wrapAnsi16m = (offset = 0) => (red, green, blue) => `\x1B[${38 + offset};2;${red};${green};${blue}m`;
var styles = {
  modifier: {
    reset: [0, 0],
    // 21 isn't widely supported and 22 does the same thing
    bold: [1, 22],
    dim: [2, 22],
    italic: [3, 23],
    underline: [4, 24],
    overline: [53, 55],
    inverse: [7, 27],
    hidden: [8, 28],
    strikethrough: [9, 29]
  },
  color: {
    black: [30, 39],
    red: [31, 39],
    green: [32, 39],
    yellow: [33, 39],
    blue: [34, 39],
    magenta: [35, 39],
    cyan: [36, 39],
    white: [37, 39],
    // Bright color
    blackBright: [90, 39],
    gray: [90, 39],
    // Alias of `blackBright`
    grey: [90, 39],
    // Alias of `blackBright`
    redBright: [91, 39],
    greenBright: [92, 39],
    yellowBright: [93, 39],
    blueBright: [94, 39],
    magentaBright: [95, 39],
    cyanBright: [96, 39],
    whiteBright: [97, 39]
  },
  bgColor: {
    bgBlack: [40, 49],
    bgRed: [41, 49],
    bgGreen: [42, 49],
    bgYellow: [43, 49],
    bgBlue: [44, 49],
    bgMagenta: [45, 49],
    bgCyan: [46, 49],
    bgWhite: [47, 49],
    // Bright color
    bgBlackBright: [100, 49],
    bgGray: [100, 49],
    // Alias of `bgBlackBright`
    bgGrey: [100, 49],
    // Alias of `bgBlackBright`
    bgRedBright: [101, 49],
    bgGreenBright: [102, 49],
    bgYellowBright: [103, 49],
    bgBlueBright: [104, 49],
    bgMagentaBright: [105, 49],
    bgCyanBright: [106, 49],
    bgWhiteBright: [107, 49]
  }
};
var modifierNames = Object.keys(styles.modifier);
var foregroundColorNames = Object.keys(styles.color);
var backgroundColorNames = Object.keys(styles.bgColor);
var colorNames = [...foregroundColorNames, ...backgroundColorNames];
function assembleStyles() {
  const codes = /* @__PURE__ */ new Map();
  for (const [groupName, group] of Object.entries(styles)) {
    for (const [styleName, style] of Object.entries(group)) {
      styles[styleName] = {
        open: `\x1B[${style[0]}m`,
        close: `\x1B[${style[1]}m`
      };
      group[styleName] = styles[styleName];
      codes.set(style[0], style[1]);
    }
    Object.defineProperty(styles, groupName, {
      value: group,
      enumerable: false
    });
  }
  Object.defineProperty(styles, "codes", {
    value: codes,
    enumerable: false
  });
  styles.color.close = "\x1B[39m";
  styles.bgColor.close = "\x1B[49m";
  styles.color.ansi = wrapAnsi16();
  styles.color.ansi256 = wrapAnsi256();
  styles.color.ansi16m = wrapAnsi16m();
  styles.bgColor.ansi = wrapAnsi16(ANSI_BACKGROUND_OFFSET);
  styles.bgColor.ansi256 = wrapAnsi256(ANSI_BACKGROUND_OFFSET);
  styles.bgColor.ansi16m = wrapAnsi16m(ANSI_BACKGROUND_OFFSET);
  Object.defineProperties(styles, {
    rgbToAnsi256: {
      value(red, green, blue) {
        if (red === green && green === blue) {
          if (red < 8) {
            return 16;
          }
          if (red > 248) {
            return 231;
          }
          return Math.round((red - 8) / 247 * 24) + 232;
        }
        return 16 + 36 * Math.round(red / 255 * 5) + 6 * Math.round(green / 255 * 5) + Math.round(blue / 255 * 5);
      },
      enumerable: false
    },
    hexToRgb: {
      value(hex) {
        const matches2 = /[a-f\d]{6}|[a-f\d]{3}/i.exec(hex.toString(16));
        if (!matches2) {
          return [0, 0, 0];
        }
        let [colorString] = matches2;
        if (colorString.length === 3) {
          colorString = [...colorString].map((character) => character + character).join("");
        }
        const integer = Number.parseInt(colorString, 16);
        return [
          /* eslint-disable no-bitwise */
          integer >> 16 & 255,
          integer >> 8 & 255,
          integer & 255
          /* eslint-enable no-bitwise */
        ];
      },
      enumerable: false
    },
    hexToAnsi256: {
      value: (hex) => styles.rgbToAnsi256(...styles.hexToRgb(hex)),
      enumerable: false
    },
    ansi256ToAnsi: {
      value(code) {
        if (code < 8) {
          return 30 + code;
        }
        if (code < 16) {
          return 90 + (code - 8);
        }
        let red;
        let green;
        let blue;
        if (code >= 232) {
          red = ((code - 232) * 10 + 8) / 255;
          green = red;
          blue = red;
        } else {
          code -= 16;
          const remainder = code % 36;
          red = Math.floor(code / 36) / 5;
          green = Math.floor(remainder / 6) / 5;
          blue = remainder % 6 / 5;
        }
        const value = Math.max(red, green, blue) * 2;
        if (value === 0) {
          return 30;
        }
        let result = 30 + (Math.round(blue) << 2 | Math.round(green) << 1 | Math.round(red));
        if (value === 2) {
          result += 60;
        }
        return result;
      },
      enumerable: false
    },
    rgbToAnsi: {
      value: (red, green, blue) => styles.ansi256ToAnsi(styles.rgbToAnsi256(red, green, blue)),
      enumerable: false
    },
    hexToAnsi: {
      value: (hex) => styles.ansi256ToAnsi(styles.hexToAnsi256(hex)),
      enumerable: false
    }
  });
  return styles;
}
var ansiStyles = assembleStyles();
var ansi_styles_default = ansiStyles;

// ../../node_modules/.pnpm/chalk@5.6.2/node_modules/chalk/source/vendor/supports-color/index.js
init_esm_shims();
import process2 from "process";
import os from "os";
import tty from "tty";
function hasFlag(flag, argv = globalThis.Deno ? globalThis.Deno.args : process2.argv) {
  const prefix2 = flag.startsWith("-") ? "" : flag.length === 1 ? "-" : "--";
  const position = argv.indexOf(prefix2 + flag);
  const terminatorPosition = argv.indexOf("--");
  return position !== -1 && (terminatorPosition === -1 || position < terminatorPosition);
}
var { env } = process2;
var flagForceColor;
if (hasFlag("no-color") || hasFlag("no-colors") || hasFlag("color=false") || hasFlag("color=never")) {
  flagForceColor = 0;
} else if (hasFlag("color") || hasFlag("colors") || hasFlag("color=true") || hasFlag("color=always")) {
  flagForceColor = 1;
}
function envForceColor() {
  if ("FORCE_COLOR" in env) {
    if (env.FORCE_COLOR === "true") {
      return 1;
    }
    if (env.FORCE_COLOR === "false") {
      return 0;
    }
    return env.FORCE_COLOR.length === 0 ? 1 : Math.min(Number.parseInt(env.FORCE_COLOR, 10), 3);
  }
}
function translateLevel(level) {
  if (level === 0) {
    return false;
  }
  return {
    level,
    hasBasic: true,
    has256: level >= 2,
    has16m: level >= 3
  };
}
function _supportsColor(haveStream, { streamIsTTY, sniffFlags = true } = {}) {
  const noFlagForceColor = envForceColor();
  if (noFlagForceColor !== void 0) {
    flagForceColor = noFlagForceColor;
  }
  const forceColor = sniffFlags ? flagForceColor : noFlagForceColor;
  if (forceColor === 0) {
    return 0;
  }
  if (sniffFlags) {
    if (hasFlag("color=16m") || hasFlag("color=full") || hasFlag("color=truecolor")) {
      return 3;
    }
    if (hasFlag("color=256")) {
      return 2;
    }
  }
  if ("TF_BUILD" in env && "AGENT_NAME" in env) {
    return 1;
  }
  if (haveStream && !streamIsTTY && forceColor === void 0) {
    return 0;
  }
  const min = forceColor || 0;
  if (env.TERM === "dumb") {
    return min;
  }
  if (process2.platform === "win32") {
    const osRelease = os.release().split(".");
    if (Number(osRelease[0]) >= 10 && Number(osRelease[2]) >= 10586) {
      return Number(osRelease[2]) >= 14931 ? 3 : 2;
    }
    return 1;
  }
  if ("CI" in env) {
    if (["GITHUB_ACTIONS", "GITEA_ACTIONS", "CIRCLECI"].some((key) => key in env)) {
      return 3;
    }
    if (["TRAVIS", "APPVEYOR", "GITLAB_CI", "BUILDKITE", "DRONE"].some((sign) => sign in env) || env.CI_NAME === "codeship") {
      return 1;
    }
    return min;
  }
  if ("TEAMCITY_VERSION" in env) {
    return /^(9\.(0*[1-9]\d*)\.|\d{2,}\.)/.test(env.TEAMCITY_VERSION) ? 1 : 0;
  }
  if (env.COLORTERM === "truecolor") {
    return 3;
  }
  if (env.TERM === "xterm-kitty") {
    return 3;
  }
  if (env.TERM === "xterm-ghostty") {
    return 3;
  }
  if (env.TERM === "wezterm") {
    return 3;
  }
  if ("TERM_PROGRAM" in env) {
    const version = Number.parseInt((env.TERM_PROGRAM_VERSION || "").split(".")[0], 10);
    switch (env.TERM_PROGRAM) {
      case "iTerm.app": {
        return version >= 3 ? 3 : 2;
      }
      case "Apple_Terminal": {
        return 2;
      }
    }
  }
  if (/-256(color)?$/i.test(env.TERM)) {
    return 2;
  }
  if (/^screen|^xterm|^vt100|^vt220|^rxvt|color|ansi|cygwin|linux/i.test(env.TERM)) {
    return 1;
  }
  if ("COLORTERM" in env) {
    return 1;
  }
  return min;
}
function createSupportsColor(stream, options = {}) {
  const level = _supportsColor(stream, {
    streamIsTTY: stream && stream.isTTY,
    ...options
  });
  return translateLevel(level);
}
var supportsColor = {
  stdout: createSupportsColor({ isTTY: tty.isatty(1) }),
  stderr: createSupportsColor({ isTTY: tty.isatty(2) })
};
var supports_color_default = supportsColor;

// ../../node_modules/.pnpm/chalk@5.6.2/node_modules/chalk/source/utilities.js
init_esm_shims();
function stringReplaceAll(string, substring, replacer) {
  let index = string.indexOf(substring);
  if (index === -1) {
    return string;
  }
  const substringLength = substring.length;
  let endIndex = 0;
  let returnValue = "";
  do {
    returnValue += string.slice(endIndex, index) + substring + replacer;
    endIndex = index + substringLength;
    index = string.indexOf(substring, endIndex);
  } while (index !== -1);
  returnValue += string.slice(endIndex);
  return returnValue;
}
function stringEncaseCRLFWithFirstIndex(string, prefix2, postfix, index) {
  let endIndex = 0;
  let returnValue = "";
  do {
    const gotCR = string[index - 1] === "\r";
    returnValue += string.slice(endIndex, gotCR ? index - 1 : index) + prefix2 + (gotCR ? "\r\n" : "\n") + postfix;
    endIndex = index + 1;
    index = string.indexOf("\n", endIndex);
  } while (index !== -1);
  returnValue += string.slice(endIndex);
  return returnValue;
}

// ../../node_modules/.pnpm/chalk@5.6.2/node_modules/chalk/source/index.js
var { stdout: stdoutColor, stderr: stderrColor } = supports_color_default;
var GENERATOR = /* @__PURE__ */ Symbol("GENERATOR");
var STYLER = /* @__PURE__ */ Symbol("STYLER");
var IS_EMPTY = /* @__PURE__ */ Symbol("IS_EMPTY");
var levelMapping = [
  "ansi",
  "ansi",
  "ansi256",
  "ansi16m"
];
var styles2 = /* @__PURE__ */ Object.create(null);
var applyOptions = (object, options = {}) => {
  if (options.level && !(Number.isInteger(options.level) && options.level >= 0 && options.level <= 3)) {
    throw new Error("The `level` option should be an integer from 0 to 3");
  }
  const colorLevel = stdoutColor ? stdoutColor.level : 0;
  object.level = options.level === void 0 ? colorLevel : options.level;
};
var chalkFactory = (options) => {
  const chalk2 = (...strings) => strings.join(" ");
  applyOptions(chalk2, options);
  Object.setPrototypeOf(chalk2, createChalk.prototype);
  return chalk2;
};
function createChalk(options) {
  return chalkFactory(options);
}
Object.setPrototypeOf(createChalk.prototype, Function.prototype);
for (const [styleName, style] of Object.entries(ansi_styles_default)) {
  styles2[styleName] = {
    get() {
      const builder = createBuilder(this, createStyler(style.open, style.close, this[STYLER]), this[IS_EMPTY]);
      Object.defineProperty(this, styleName, { value: builder });
      return builder;
    }
  };
}
styles2.visible = {
  get() {
    const builder = createBuilder(this, this[STYLER], true);
    Object.defineProperty(this, "visible", { value: builder });
    return builder;
  }
};
var getModelAnsi = (model, level, type, ...arguments_) => {
  if (model === "rgb") {
    if (level === "ansi16m") {
      return ansi_styles_default[type].ansi16m(...arguments_);
    }
    if (level === "ansi256") {
      return ansi_styles_default[type].ansi256(ansi_styles_default.rgbToAnsi256(...arguments_));
    }
    return ansi_styles_default[type].ansi(ansi_styles_default.rgbToAnsi(...arguments_));
  }
  if (model === "hex") {
    return getModelAnsi("rgb", level, type, ...ansi_styles_default.hexToRgb(...arguments_));
  }
  return ansi_styles_default[type][model](...arguments_);
};
var usedModels = ["rgb", "hex", "ansi256"];
for (const model of usedModels) {
  styles2[model] = {
    get() {
      const { level } = this;
      return function(...arguments_) {
        const styler = createStyler(getModelAnsi(model, levelMapping[level], "color", ...arguments_), ansi_styles_default.color.close, this[STYLER]);
        return createBuilder(this, styler, this[IS_EMPTY]);
      };
    }
  };
  const bgModel = "bg" + model[0].toUpperCase() + model.slice(1);
  styles2[bgModel] = {
    get() {
      const { level } = this;
      return function(...arguments_) {
        const styler = createStyler(getModelAnsi(model, levelMapping[level], "bgColor", ...arguments_), ansi_styles_default.bgColor.close, this[STYLER]);
        return createBuilder(this, styler, this[IS_EMPTY]);
      };
    }
  };
}
var proto = Object.defineProperties(() => {
}, {
  ...styles2,
  level: {
    enumerable: true,
    get() {
      return this[GENERATOR].level;
    },
    set(level) {
      this[GENERATOR].level = level;
    }
  }
});
var createStyler = (open, close, parent2) => {
  let openAll;
  let closeAll;
  if (parent2 === void 0) {
    openAll = open;
    closeAll = close;
  } else {
    openAll = parent2.openAll + open;
    closeAll = close + parent2.closeAll;
  }
  return {
    open,
    close,
    openAll,
    closeAll,
    parent: parent2
  };
};
var createBuilder = (self, _styler, _isEmpty) => {
  const builder = (...arguments_) => applyStyle(builder, arguments_.length === 1 ? "" + arguments_[0] : arguments_.join(" "));
  Object.setPrototypeOf(builder, proto);
  builder[GENERATOR] = self;
  builder[STYLER] = _styler;
  builder[IS_EMPTY] = _isEmpty;
  return builder;
};
var applyStyle = (self, string) => {
  if (self.level <= 0 || !string) {
    return self[IS_EMPTY] ? "" : string;
  }
  let styler = self[STYLER];
  if (styler === void 0) {
    return string;
  }
  const { openAll, closeAll } = styler;
  if (string.includes("\x1B")) {
    while (styler !== void 0) {
      string = stringReplaceAll(string, styler.close, styler.open);
      styler = styler.parent;
    }
  }
  const lfIndex = string.indexOf("\n");
  if (lfIndex !== -1) {
    string = stringEncaseCRLFWithFirstIndex(string, closeAll, openAll, lfIndex);
  }
  return openAll + string + closeAll;
};
Object.defineProperties(createChalk.prototype, styles2);
var chalk = createChalk();
var chalkStderr = createChalk({ level: stderrColor ? stderrColor.level : 0 });
var source_default = chalk;

// ../../node_modules/.pnpm/@wdio+logger@9.18.0/node_modules/@wdio/logger/build/index.js
var import_loglevel_plugin_prefix = __toESM(require_loglevel_plugin_prefix(), 1);

// ../../node_modules/.pnpm/strip-ansi@7.2.0/node_modules/strip-ansi/index.js
init_esm_shims();

// ../../node_modules/.pnpm/ansi-regex@6.2.2/node_modules/ansi-regex/index.js
init_esm_shims();
function ansiRegex({ onlyFirst = false } = {}) {
  const ST = "(?:\\u0007|\\u001B\\u005C|\\u009C)";
  const osc = `(?:\\u001B\\][\\s\\S]*?${ST})`;
  const csi = "[\\u001B\\u009B][[\\]()#;?]*(?:\\d{1,4}(?:[;:]\\d{0,4})*)?[\\dA-PR-TZcf-nq-uy=><~]";
  const pattern = `${osc}|${csi}`;
  return new RegExp(pattern, onlyFirst ? void 0 : "g");
}

// ../../node_modules/.pnpm/strip-ansi@7.2.0/node_modules/strip-ansi/index.js
var regex = ansiRegex();
function stripAnsi(string) {
  if (typeof string !== "string") {
    throw new TypeError(`Expected a \`string\`, got \`${typeof string}\``);
  }
  if (!string.includes("\x1B") && !string.includes("\x9B")) {
    return string;
  }
  return string.replace(regex, "");
}

// ../../node_modules/.pnpm/@wdio+logger@9.18.0/node_modules/@wdio/logger/build/index.js
var SENSITIVE_DATA_REPLACER = "**MASKED**";
var skipError = (aFunction) => {
  try {
    return aFunction();
  } catch {
    return void 0;
  }
};
var parseMaskingPatterns = (maskingRegexString) => {
  if (typeof maskingRegexString !== "string") {
    return void 0;
  }
  const regexStrings = maskingRegexString?.split(/,\s*/).filter((regexStr) => regexStr.trim() !== "");
  return regexStrings?.map((regexStr) => {
    const regexParts = regexStr.match(/^\/(.*?)\/([gimsuy]*)$/);
    if (!regexParts && (0, import_safe_regex2.default)(regexStr)) {
      return skipError(() => new RegExp(regexStr));
    }
    if (regexParts?.[1] && (0, import_safe_regex2.default)(regexParts[1])) {
      return skipError(() => regexParts[2] ? new RegExp(regexParts[1], regexParts[2]) : new RegExp(regexParts[1]));
    }
    return void 0;
  }).filter((regex2) => regex2 !== void 0);
};
var mask = (text, maskingPatterns) => {
  if (!maskingPatterns || typeof text !== "string") {
    return text;
  }
  const endsWithNewline = text.endsWith("\n");
  let maskedText = text;
  maskingPatterns.forEach((maskingRegex) => {
    maskedText = maskedText.replace(maskingRegex, (fullMatch, ...capturedGroupsAndMore) => {
      const capturedGroups = capturedGroupsAndMore.slice(0, capturedGroupsAndMore.length - 2);
      if (capturedGroups.length === 0) {
        return SENSITIVE_DATA_REPLACER;
      }
      let matchedMaskedText = fullMatch;
      capturedGroups.forEach((group) => {
        matchedMaskedText = matchedMaskedText.replace(group, SENSITIVE_DATA_REPLACER);
      });
      return matchedMaskedText;
    });
  });
  if (endsWithNewline && !maskedText.endsWith("\n")) {
    maskedText += "\n";
  }
  return maskedText;
};
import_loglevel_plugin_prefix.default.reg(import_loglevel.default);
var DEFAULT_LEVEL = process.env.WDIO_DEBUG ? "trace" : "info";
var COLORS = {
  error: "red",
  warn: "yellow",
  info: "cyanBright",
  debug: "green",
  trace: "cyan",
  progress: "magenta"
};
var matches = {
  COMMAND: "COMMAND",
  BIDICOMMAND: "BIDI COMMAND",
  DATA: "DATA",
  RESULT: "RESULT",
  BIDIRESULT: "BIDI RESULT"
};
var SERIALIZERS = [{
  /**
   * display error stack
   */
  matches: (err) => err instanceof Error,
  serialize: (err) => err.stack
}, {
  /**
   * color commands blue
   */
  matches: (log22) => log22 === matches.COMMAND || log22 === matches.BIDICOMMAND,
  serialize: (log22) => source_default.magenta(log22)
}, {
  /**
   * color data yellow
   */
  matches: (log22) => log22 === matches.DATA,
  serialize: (log22) => source_default.yellow(log22)
}, {
  /**
   * color result cyan
   */
  matches: (log22) => log22 === matches.RESULT || log22 === matches.BIDIRESULT,
  serialize: (log22) => source_default.cyan(log22)
}];
var loggers = import_loglevel.default.getLoggers();
var logLevelsConfig = {};
var maskingPatternsConfig = {};
var logCache = /* @__PURE__ */ new Set();
var logFile;
var originalFactory = import_loglevel.default.methodFactory;
var wdioLoggerMethodFactory = (wdioLogger) => function(methodName, logLevel, loggerName) {
  const rawMethod = originalFactory(methodName, logLevel, loggerName);
  return (...args) => {
    if (!logFile && process.env.WDIO_LOG_PATH) {
      logFile = fs.createWriteStream(process.env.WDIO_LOG_PATH);
    }
    const match = Object.values(matches).filter((x) => args[0].endsWith(`: ${x}`))[0];
    if (match) {
      const prefixStr = args.shift().slice(0, -match.length - 1);
      args.unshift(prefixStr, match);
    }
    args = args.map((arg) => {
      for (const s of SERIALIZERS) {
        if (s.matches(arg)) {
          return s.serialize(arg);
        }
      }
      return arg;
    });
    const unmaskedLogText = stripAnsi(`${util.format.apply(this, args)}
`);
    const maskedLogText = mask(unmaskedLogText, wdioLogger.maskingPatterns);
    if (logFile && logFile.writable) {
      if (logCache.size) {
        logCache.forEach((log22) => {
          if (logFile) {
            logFile.write(log22);
          }
        });
        logCache.clear();
      }
      if (!logsContainInitPackageError(unmaskedLogText)) {
        return logFile.write(maskedLogText);
      }
      logFile.write(maskedLogText);
    }
    logCache.add(maskedLogText);
    if (maskedLogText === unmaskedLogText) {
      rawMethod(...args);
    } else {
      rawMethod(maskedLogText.replace(/\n$/, ""));
    }
  };
};
var progress = function(data) {
  if (process.stdout.isTTY && this.getLevel() <= import_loglevel.default.levels.INFO) {
    const level = "progress";
    const timestampFormatter = source_default.gray((/* @__PURE__ */ new Date()).toISOString());
    const levelFormatter = source_default[COLORS[level]](level.toUpperCase());
    const nameFormatter = source_default.whiteBright(this.name);
    const _data = data.length > 0 ? `${timestampFormatter} ${levelFormatter} ${nameFormatter}: ${data}` : "\r\x1B[K\x1B[?25h";
    process.stdout.write("\x1B[?25l");
    process.stdout.write(`${_data}\r`);
  }
};
function getLogger(name) {
  if (loggers[name]) {
    return loggers[name];
  }
  let logLevel = process.env.WDIO_LOG_LEVEL || DEFAULT_LEVEL;
  const logLevelName = getLogLevelName(name);
  if (logLevelsConfig[logLevelName]) {
    logLevel = logLevelsConfig[logLevelName];
  }
  loggers[name] = import_loglevel.default.getLogger(name);
  const logger = loggers[name];
  logger.setLevel(logLevel);
  logger.maskingPatterns = maskingPatternsConfig[name] ?? parseMaskingPatterns(process.env.WDIO_LOG_MASKING_PATTERNS);
  logger.progress = progress;
  logger.methodFactory = wdioLoggerMethodFactory(logger);
  import_loglevel_plugin_prefix.default.apply(logger, {
    template: "%t %l %n:",
    timestampFormatter: (date) => source_default.gray(date.toISOString()),
    levelFormatter: (level) => source_default[COLORS[level]](level.toUpperCase()),
    nameFormatter: (name2) => source_default.whiteBright(name2)
  });
  return logger;
}
getLogger.waitForBuffer = async () => new Promise((resolve2) => {
  if (logFile && Array.isArray(logFile.writableBuffer) && logFile.writableBuffer.length !== 0) {
    return setTimeout(async () => {
      await getLogger.waitForBuffer();
      resolve2();
    }, 20);
  }
  resolve2();
});
getLogger.setLevel = (name, level) => loggers[name].setLevel(level);
getLogger.clearLogger = () => {
  if (logFile) {
    logFile.end();
  }
  logFile = null;
};
getLogger.setLogLevelsConfig = (logLevels = {}, wdioLogLevel = DEFAULT_LEVEL) => {
  if (process.env.WDIO_LOG_LEVEL === void 0) {
    process.env.WDIO_LOG_LEVEL = wdioLogLevel;
  }
  logLevelsConfig = {};
  Object.entries(logLevels).forEach(([logName, logLevel]) => {
    const logLevelName = getLogLevelName(logName);
    logLevelsConfig[logLevelName] = logLevel;
  });
  Object.keys(loggers).forEach((logName) => {
    const logLevelName = getLogLevelName(logName);
    const logLevel = typeof logLevelsConfig[logLevelName] !== "undefined" ? logLevelsConfig[logLevelName] : process.env.WDIO_LOG_LEVEL;
    loggers[logName].setLevel(logLevel);
  });
};
getLogger.setMaskingPatterns = (pattern) => {
  if (typeof pattern === "string") {
    if (process.env.WDIO_LOG_MASKING_PATTERNS === void 0) {
      process.env.WDIO_LOG_MASKING_PATTERNS = pattern;
    }
  } else if (typeof pattern === "object") {
    maskingPatternsConfig = Object.entries(pattern).reduce((acc, [logName, maskingPatternsString]) => {
      acc[logName] = parseMaskingPatterns(maskingPatternsString);
      return acc;
    }, maskingPatternsConfig);
  } else {
    throw new Error(`Invalid pattern property, expected \`string\` or \`Record<string, string>\` but received \`${typeof pattern}\``);
  }
  Object.keys(loggers).forEach((logName) => {
    const maskingPatterns = maskingPatternsConfig[logName] ?? parseMaskingPatterns(process.env.WDIO_LOG_MASKING_PATTERNS);
    loggers[logName].maskingPatterns = maskingPatterns;
  });
};
var getLogLevelName = (logName) => logName.split(":").shift();
function logsContainInitPackageError(logText) {
  return ERROR_LOG_VALIDATOR.every((pattern) => logText.includes(pattern));
}
var ERROR_LOG_VALIDATOR = [
  "Couldn't find plugin",
  "neither as wdio scoped package",
  "nor as community package",
  "Please make sure you have it installed"
];

// src/service.ts
import { fileURLToPath as fileURLToPath3 } from "url";
import ObsidianLauncher from "obsidian-launcher";

// src/browserCommands.ts
init_esm_shims();

// src/pageobjects/obsidianPage.ts
init_esm_shims();
import * as path3 from "path";
import * as fsAsync2 from "fs/promises";
import * as crypto2 from "crypto";
import * as semver from "semver";
import { fileURLToPath as fileURLToPath2 } from "url";
import _2 from "lodash";

// src/types.ts
init_esm_shims();
var OBSIDIAN_CAPABILITY_KEY = "wdio:obsidianOptions";

// src/pageobjects/basePage.ts
init_esm_shims();

// ../../node_modules/.pnpm/@wdio+globals@9.17.0_expect-webdriverio@5.6.5_webdriverio@9.18.1/node_modules/@wdio/globals/build/index.js
init_esm_shims();
var globals = globalThis._wdioGlobals = globalThis._wdioGlobals || /* @__PURE__ */ new Map();
var GLOBALS_ERROR_MESSAGE = `No browser instance registered. Don't import @wdio/globals outside of the WDIO testrunner context. Or you have two two different "@wdio/globals" packages installed.`;
function proxyHandler(key) {
  return {
    get: (self, prop) => {
      if (!globals.has(key)) {
        throw new Error(GLOBALS_ERROR_MESSAGE);
      }
      const receiver = globals.get(key);
      const field = receiver[prop];
      return typeof field === "function" ? field.bind(receiver) : field;
    }
  };
}
var browser = new Proxy(
  class Browser {
  },
  proxyHandler("browser")
);
var driver = new Proxy(
  class Browser2 {
  },
  proxyHandler("driver")
);
var multiremotebrowser = new Proxy(
  class Browser3 {
  },
  proxyHandler("multiremotebrowser")
);
var expect = (...args) => {
  if (!globals.has("expect")) {
    throw new Error(GLOBALS_ERROR_MESSAGE);
  }
  return globals.get("expect")(...args);
};
var ASYNC_MATCHERS = [
  "any",
  "anything",
  "arrayContaining",
  "objectContaining",
  "stringContaining",
  "stringMatching"
];
for (const matcher of ASYNC_MATCHERS) {
  expect[matcher] = (...args) => {
    if (!globals.has("expect")) {
      throw new Error(GLOBALS_ERROR_MESSAGE);
    }
    return globals.get("expect")[matcher](...args);
  };
}
expect.not = ASYNC_MATCHERS.reduce((acc, matcher) => {
  acc[matcher] = (...args) => {
    if (!globals.has("expect")) {
      throw new Error(GLOBALS_ERROR_MESSAGE);
    }
    return globals.get("expect").not[matcher](...args);
  };
  return acc;
}, {});
expect.extend = (...args) => {
  if (!globals.has("expect")) {
    throw new Error(GLOBALS_ERROR_MESSAGE);
  }
  const expect2 = globals.get("expect");
  return expect2.extend(...args);
};

// src/pageobjects/basePage.ts
var BasePage = class {
  constructor(browser2) {
    this._browser = browser2;
  }
  /**
   * Returns the browser instance.
   * @hidden
   */
  get browser() {
    return this._browser ?? browser;
  }
};

// src/utils.ts
init_esm_shims();
import path2 from "path";
import crypto from "crypto";
import fsAsync from "fs/promises";
import * as tar from "tar";
import _ from "lodash";
function quote(input) {
  return `'${input.replace(/'/g, "'\\''")}'`;
}
async function fileExists(path5) {
  try {
    await fsAsync.access(path5);
    return true;
  } catch {
    return false;
  }
}
function getAppiumOptions(cap) {
  let result = _.pickBy(cap, (v, k) => k.startsWith("appium:") && k != "appium:options");
  result = _.mapKeys(result, (v, k) => k.slice(7));
  return { ...result, ...cap["appium:options"] };
}
function isAppium(cap) {
  return getAppiumOptions(cap).automationName?.toLocaleLowerCase() === "uiautomator2";
}
async function appiumUploadFiles(browser2, opts) {
  let {
    src,
    dest,
    files = [src],
    chunkSize = 2 * 1024 * 1024
  } = opts;
  src = path2.resolve(src);
  dest = path2.posix.normalize(dest).replace(/\/$/, "");
  files = files.map((f) => path2.relative(src, f) || ".");
  if (files.length == 0) return;
  const tmpDir = "/data/local/tmp";
  const slug = crypto.randomBytes(10).toString("base64url").replace(/[-_]/g, "0");
  const stream = tar.create({ C: src, gzip: { level: 2 } }, files);
  let buffers = [];
  let bufferSize = 0;
  let i = 0;
  for await (const chunk of stream) {
    if (bufferSize + chunk.length > chunkSize && bufferSize > 0) {
      const data2 = Buffer.concat(buffers).toString("base64");
      await browser2.pushFile(`${tmpDir}/${slug}-${String(i).padStart(6, "0")}.tar`, data2);
      i++;
      buffers = [];
      bufferSize = 0;
    }
    buffers.push(chunk);
    bufferSize += chunk.length;
  }
  const data = Buffer.concat(buffers).toString("base64");
  await browser2.pushFile(`${tmpDir}/${slug}-${String(i).padStart(6, "0")}.tar`, data);
  await browser2.execute("mobile: shell", { command: "sh", args: ["-c", quote(`
        mkdir -p ${quote(dest)};
        cat ${tmpDir}/${slug}-*.tar | tar -xz -C ${quote(dest)};
        rm ${tmpDir}/${slug}-*.tar;
    `)] });
}
async function appiumDownloadFile(browser2, src, dest) {
  src = path2.posix.normalize(src);
  dest = path2.resolve(dest);
  const content = Buffer.from(await browser2.pullFile(src), "base64");
  await fsAsync.writeFile(dest, content);
}
async function navigateAndWait(browser2, func) {
  await browser2.execute(() => {
    window._waitingForNavigation = true;
  });
  await browser2.execute(func);
  await browser2.waitUntil(() => browser2.execute(() => !("_waitingForNavigation" in window)));
}
function normalizePath(p) {
  p = p.replace(/([\\/])+/g, "/").replace(/(^\/+|\/+$)/g, "");
  if (p == "") {
    p = "/";
  }
  p = p.replace(/\u00A0|\u202F/g, " ");
  p = p.normalize("NFC");
  return p;
}
function isHidden(file) {
  return file.split("/").some((p) => p.startsWith("."));
}
function isText(file) {
  return [".md", ".json", ".txt", ".js"].includes(path2.extname(file).toLocaleLowerCase());
}

// src/pageobjects/obsidianPage.ts
var ObsidianPage = class extends BasePage {
  getObsidianCapabilities() {
    return this.browser.requestedCapabilities[OBSIDIAN_CAPABILITY_KEY];
  }
  /**
   * Returns the path to the vault opened in Obsidian.
   * 
   * wdio-obsidian-service copies your vault before running tests, so this is the path to the temporary copy.
   */
  getVaultPath() {
    const obsidianOptions = this.getObsidianCapabilities();
    if (obsidianOptions.vaultCopy === void 0) {
      throw Error("No vault open, set vault in wdio.conf.mts or use reloadObsidian to open a vault dynamically.");
    }
    return obsidianOptions.uploadedVault ?? obsidianOptions.vaultCopy;
  }
  /** 
   * Run an Obsidian CLI command.
   * 
   * The Obsidian CLI Only works on Obsidian >=1.12.0 with installer >=1.11.7
   * See https://help.obsidian.md/cli
   */
  async runObsidianCli(args) {
    const { appVersion, installerVersion } = this.getObsidianCapabilities();
    if (semver.lt(appVersion, "1.12.0") || semver.lt(installerVersion, "1.11.7")) {
      throw Error(`Obsidian CLI only works on app >=1.12.0 and installer >=1.11.7`);
    }
    if (isAppium(this.browser.requestedCapabilities)) {
      throw Error(`Obsidian CLI only works on desktop`);
    }
    const stdout = await this.browser.execute((args2) => window.handleCli(args2), args);
    return { stdout, stderr: "" };
  }
  /**
   * Return the Obsidian config dir (just ".obsidian" unless you changed the config dir name in settings).
   */
  async getConfigDir() {
    return await this.browser.executeObsidian(({ app }) => app.vault.configDir);
  }
  /**
   * Returns the Obsidian Platform object. Useful for skipping tests based on whether you are running in desktop or
   * mobile, or based on OS.
   */
  async getPlatform() {
    const obsidianOptions = this.getObsidianCapabilities();
    const appium = isAppium(this.browser.requestedCapabilities);
    const emulateMobile = obsidianOptions.emulateMobile;
    const platform = {
      isDesktop: !(appium || emulateMobile),
      isMobile: appium || emulateMobile,
      isDesktopApp: !appium,
      isMobileApp: appium,
      isIosApp: false,
      // iOS is not supported
      isAndroidApp: appium,
      isMacOS: !appium && process.platform == "darwin",
      isWin: !appium && process.platform == "win32",
      isLinux: appium || process.platform == "linux",
      isSafari: false
      // iOS is not supported
    };
    if (obsidianOptions.vault !== void 0) {
      const obsidianPlatform = await this.browser.executeObsidian(({ obsidian }) => {
        const p = obsidian.Platform;
        return {
          isMobileApp: p.isMobileApp,
          isIosApp: p.isIosApp,
          isAndroidApp: p.isAndroidApp,
          isPhone: p.isPhone,
          isTablet: p.isTablet,
          isMacOS: p.isMacOS,
          isWin: p.isWin,
          isLinux: p.isLinux,
          isSafari: p.isSafari
        };
      });
      Object.assign(platform, _2.pickBy(obsidianPlatform, (x) => x != null));
    } else {
      const [width, height] = await this.browser.execute(() => [window.innerWidth, window.innerHeight]);
      let isTablet = false;
      let isPhone = false;
      if (appium || emulateMobile) {
        isTablet = width >= 600 && height >= 600;
        isPhone = !isTablet;
      }
      Object.assign(platform, { isTablet, isPhone });
    }
    return platform;
  }
  /**
   * Enables a plugin by ID
   */
  async enablePlugin(pluginId) {
    await this.browser.executeObsidian(
      async ({ app }, pluginId2) => await app.plugins.enablePluginAndSave(pluginId2),
      pluginId
    );
  }
  /**
   * Disables a plugin by ID
   */
  async disablePlugin(pluginId) {
    await this.browser.executeObsidian(
      async ({ app }, pluginId2) => await app.plugins.disablePluginAndSave(pluginId2),
      pluginId
    );
  }
  /**
   * Sets the theme. Pass "default" to reset to the Obsidian theme.
   */
  async setTheme(themeName) {
    themeName = themeName == "default" ? "" : themeName;
    await this.browser.executeObsidian(
      async ({ app }, themeName2) => await app.customCss.setTheme(themeName2),
      themeName
    );
  }
  /**
   * Opens a file in a new tab.
   */
  async openFile(path5) {
    await this.browser.executeObsidian(async ({ app, obsidian }, path6) => {
      const file = app.vault.getAbstractFileByPath(path6);
      if (file instanceof obsidian.TFile) {
        const leaf = app.workspace.getLeaf("tab");
        await leaf.openFile(file);
        app.workspace.setActiveLeaf(leaf, { focus: true });
      } else {
        throw Error(`No file ${path6} exists`);
      }
    }, path5);
  }
  /**
   * Loads a saved workspace layout from `.obsidian/workspaces.json` by name. Use the core "Workspaces"
   * plugin to create the layouts. You can also pass the layout object directly.
   */
  async loadWorkspaceLayout(layout) {
    if (typeof layout == "string") {
      const workspacesPath = `${await this.getConfigDir()}/workspaces.json`;
      const layoutName = layout;
      try {
        const fileContent = await this.browser.executeObsidian(async ({ app }, workspacesPath2) => {
          return await app.vault.adapter.read(workspacesPath2);
        }, workspacesPath);
        layout = JSON.parse(fileContent)?.workspaces?.[layoutName];
      } catch {
        throw new Error(`Failed to load ${workspacesPath}:${layoutName}`);
      }
      if (!layout) {
        throw new Error(`No workspace ${layoutName} found in ${workspacesPath}`);
      }
    }
    await this.browser.executeObsidian(async ({ app }, layout2) => {
      await app.workspace.changeLayout(layout2);
    }, layout);
  }
  /**
   * Return contents of a file in the vault.
   */
  async read(file) {
    return await this.browser.executeObsidian(async ({ app, obsidian }, file2) => {
      file2 = obsidian.normalizePath(file2);
      return await app.vault.adapter.read(file2);
    }, file);
  }
  /**
   * Return contents of a binary file in the vault.
   */
  async readBinary(file) {
    const b64Contents = await this.browser.executeObsidian(async ({ app, obsidian }, file2) => {
      function toB64(buffer) {
        let binary = "";
        const bytes = new Uint8Array(buffer);
        const len = bytes.byteLength;
        for (let i = 0; i < len; i++) {
          binary += String.fromCharCode(bytes[i]);
        }
        return btoa(binary);
      }
      file2 = obsidian.normalizePath(file2);
      return toB64(await app.vault.adapter.readBinary(file2));
    }, file);
    return Buffer.from(b64Contents, "base64").buffer;
  }
  /**
   * Deletes a file or folder in the vault.
   * @param file path of inside vault, e.g. "/books/leviathan-wakes.md"
   */
  async delete(file) {
    await this.browser.executeObsidian(async ({ app, obsidian }, file2, isVaultFile) => {
      file2 = obsidian.normalizePath(file2);
      if (isVaultFile) {
        const fileObj = app.vault.getAbstractFileByPath(file2);
        if (fileObj) {
          await app.vault.delete(fileObj, true);
        }
        ;
      } else {
        const stat = await app.vault.adapter.stat(file2);
        if (stat && stat.type == "folder") {
          await app.vault.adapter.rmdir(file2, true);
        } else if (stat) {
          await app.vault.adapter.remove(file2);
        }
      }
    }, file, !isHidden(file));
  }
  /**
   * Writes to a file in the vault. Creates parent directories if needed.
   * @param file path of inside vault, e.g. "/books/leviathan-wakes.md"
   * @param content content to write to the file
  */
  async write(file, content) {
    let strContent, binContent;
    if (typeof content == "string") {
      strContent = content;
    } else {
      binContent = Buffer.from(content).toString("base64");
    }
    await this.browser.executeObsidian(async ({ app, obsidian }, file2, isVaultFile, strContent2, binContent2) => {
      file2 = obsidian.normalizePath(file2);
      const parent2 = file2.split("/").slice(0, -1).join("/");
      let success = false;
      let retries = 0;
      while (!success && retries < 8) {
        if (isVaultFile) {
          if (parent2 && !app.vault.getAbstractFileByPath(parent2)) {
            await app.vault.createFolder(parent2);
          }
          const fileObj = app.vault.getAbstractFileByPath(file2);
          strContent2 = strContent2 ?? atob(binContent2);
          if (fileObj) {
            await app.vault.modify(fileObj, strContent2);
          } else {
            await app.vault.create(file2, strContent2);
          }
        } else {
          await app.vault.adapter.mkdir(parent2);
          if (strContent2) {
            await app.vault.adapter.write(file2, strContent2);
          } else {
            const buffer = Uint8Array.from(atob(binContent2), (c) => c.charCodeAt(0)).buffer;
            await app.vault.adapter.writeBinary(file2, buffer);
          }
        }
        success = !obsidian.Platform.isAndroidApp || (strContent2?.length ?? binContent2.length) === 0 || (await app.vault.adapter.stat(file2)).size > 0;
        retries++;
      }
      if (!success) {
        throw new Error(`Failed to write file ${file2}`);
      }
    }, file, !isHidden(file) && isText(file), strContent, binContent);
  }
  /**
   * Create a folder in the vault. Creates parent directories if needed.
   * @param file path of inside vault, e.g. "/books"
   */
  async mkdir(file) {
    await this.browser.executeObsidian(async ({ app, obsidian }, file2, isVaultFile) => {
      file2 = obsidian.normalizePath(file2);
      if (isVaultFile) {
        if (parent && !app.vault.getAbstractFileByPath(file2)) {
          await app.vault.createFolder(file2);
        }
      } else {
        await app.vault.adapter.mkdir(file2);
      }
    }, file, !isHidden(file));
  }
  /**
   * Updates the vault by modifying files in place without reloading Obsidian. Can be used to reset the vault back to
   * its original state or to "switch" to an entirely different vault without rebooting Obsidian
   * 
   * This will only update regular vault files, it won't touch anything under `.obsidian`, and it won't reset any
   * Obsidian config or plugin settings. But if all you need is to reset the vault files, this can be used as a faster
   * alternative to {@link ObsidianBrowserCommands.reloadObsidian | reloadObsidian}.
   * 
   * You'll often want to combine resetVault with something like this to reset your plugin's configuration as well:
   * ```ts
   * await browser.executeObsidian(async ({plugins}, settings) => {
   *     Object.assign(plugins.myPlugin.settings, settings);
   *     await plugins.myPlugin.saveSettings();
   * }, {...});
   * ```
   *
   * If no vault is passed, it resets the vault back to the oringal vault opened by the tests. You can also pass a
   * path to a different vault, and it will replace the current files with the files of that vault (similar to an
   * "rsync"). Or, instead of passing a vault path you can pass an object mapping vault file paths to file content.
   * E.g.
   * ```ts
   * obsidianPage.resetVault({
   *     'path/in/vault.md': "Hello World",
   * })
   * ```
   * 
   * You can also pass multiple vaults and objects, and they will be merged. This can be useful if you want to add a
   * few small modifications to the base vault. e.g:
   * ```ts
   * obsidianPage.resetVault('./path/to/vault', {
   *    "books/leviathan-wakes.md": "...",
   * })
   * ```
   */
  async resetVault(...vaults) {
    const obsidianOptions = this.getObsidianCapabilities();
    if (!obsidianOptions.vault) {
      const defaultVaultPath = path3.resolve(fileURLToPath2(import.meta.url), "../../default-vault");
      await this.browser.reloadObsidian({ vault: defaultVaultPath });
    }
    const configDir = await this.getConfigDir();
    vaults = vaults.length == 0 ? [obsidianOptions.vault] : vaults;
    const newVault = /* @__PURE__ */ new Map();
    for (let vault of vaults) {
      if (typeof vault == "string") {
        vault = path3.resolve(vault);
        const files = await fsAsync2.readdir(vault, { recursive: true, withFileTypes: true });
        for (const f of files) {
          const fullPath = path3.join(f.parentPath, f.name);
          const vaultPath = path3.relative(vault, fullPath).split(path3.sep).join("/");
          if (!vaultPath.startsWith(configDir + "/")) {
            if (f.isDirectory()) {
              newVault.set(vaultPath, { type: "folder" });
            } else {
              newVault.set(vaultPath, { type: "file", sourceFile: fullPath });
            }
          }
        }
      } else {
        vault = _2.mapKeys(vault, (v, k) => normalizePath(k));
        for (const [file, content] of Object.entries(vault)) {
          newVault.set(file, { type: "file", sourceContent: content });
        }
        const folders = new Set(Object.keys(vault).map((p) => path3.posix.dirname(p)).filter((p) => p !== "."));
        for (const folder of folders) {
          newVault.set(folder, { type: "folder" });
        }
      }
    }
    const currVault = new Map(await this.browser.executeObsidian(({ app }, configDir2) => {
      async function hash(data) {
        const hashBuffer = await window.crypto.subtle.digest("SHA-256", data);
        const hashArray = Array.from(new Uint8Array(hashBuffer));
        return hashArray.map((b) => b.toString(16).padStart(2, "0")).join("");
      }
      async function listRecursive(path5) {
        const result = [];
        const { folders, files } = await app.vault.adapter.list(path5);
        for (const folder of folders) {
          if (!folder.startsWith(configDir2 + "/")) {
            result.push([folder, { type: "folder" }]);
            result.push(...await listRecursive(folder));
          }
        }
        for (const file of files) {
          if (!file.startsWith(configDir2 + "/")) {
            let fileHash = app.metadataCache.getFileInfo(file)?.hash;
            if (!fileHash) {
              fileHash = await hash(await app.vault.adapter.readBinary(file));
            }
            result.push([file, { type: "file", hash: fileHash }]);
          }
        }
        return result;
      }
      return listRecursive("/");
    }, configDir));
    for (const file of [...currVault.keys()].sort().reverse()) {
      const currFileInfo = currVault.get(file);
      if (!newVault.has(file) || newVault.get(file).type != currFileInfo.type) {
        await this.delete(file);
      }
    }
    for (const [file, newFileInfo] of _2.sortBy([...newVault.entries()], 0)) {
      const currFileInfo = currVault.get(file);
      if (newFileInfo.type == "file") {
        let content = newFileInfo.sourceContent;
        if (!content) {
          content = (await fsAsync2.readFile(newFileInfo.sourceFile)).buffer;
        }
        const hash = crypto2.createHash("SHA256").update(typeof content == "string" ? content : new Uint8Array(content)).digest("hex");
        if (!currFileInfo || currFileInfo.hash != hash) {
          await this.write(file, content);
        }
      } else if (newFileInfo.type == "folder" && !currFileInfo) {
        await this.mkdir(file);
      }
    }
  }
};
var obsidianPage = new ObsidianPage();
var obsidianPage_default = obsidianPage;

// src/browserCommands.ts
var browserCommands = {
  /**
   * Wrapper around browser.execute that passes the Obsidian API to the function. The first argument to the function
   * is an object containing keys:
   * - app: Obsidian app instance
   * - obsidian: Full Obsidian API
   * - plugins: Object of all installed plugins, mapped by plugin id converted to camelCase.
   * - require: The customized require function Obsidian makes available to plugins. This is also available globally,
   *            so you can just use `require` directly instead of from {@link ExecuteObsidianArg} if you prefer.
   * 
   * The same caveats as `browser.execute` apply. The function will be stringified and then run inside Obsidian, so
   * you can't capture any local variables. Instead, pass inputs as extra arguments to the function.
   * 
   * See also: https://webdriver.io/docs/api/browser/execute
   * 
   * Example usage:
   * ```ts
   * const stats = await browser.executeObsidian(({app}, path) => {
   *     return app.vault.getMarkdownFiles().find(f => f.path == path)!.stat;
   * }, "Welcome.md");
   * ```
   * 
   * Make sure not to capture Obsidian types, instead use the provided obsidian module object. E.g. 
   * This *won't* work:
   * ```ts
   * import { FileView } from "obsidian"
   * browser.executeObsidian(({app}) => {
   *     if (leaf.view instanceof FileView) {
   *         ...
   *     }
   * })
   * ```
   * do this instead:
   * ```ts
   * browser.executeObsidian(({app, obsidian}) => {
   *     if (leaf.view instanceof obsidian.FileView) {
   *         ...
   *     }
   * })
   * ```
   */
  async executeObsidian(func, ...params) {
    const obsidianOptions = this.requestedCapabilities[OBSIDIAN_CAPABILITY_KEY];
    if (obsidianOptions.vault === void 0) {
      throw Error("No vault open, set vault in wdio.conf.mts or use reloadObsidian to open a vault dynamically.");
    }
    const result = await this.execute(`
            const require = window.wdioObsidianService().require;
            try {
                return await (
                    ${func.toString()}
                ).call(null, window.wdioObsidianService(), ...arguments);
            } catch (e) {
                if ("code" in e && typeof e.code != "number") {
                    delete e.code;
                }
                throw e;
            }
        `, ...params);
    return result;
  },
  /**
   * Executes an Obsidian command by id.
   * @param id Id of the command to run.
   */
  async executeObsidianCommand(id) {
    const result = await this.executeObsidian(
      ({ app }, id2) => app.commands.executeCommandById(id2),
      id
    );
    if (!result) {
      throw Error(`Obsidian command ${id} not found or failed.`);
    }
  },
  /**
   * Returns the Obsidian app version this test is running under.
   */
  getObsidianVersion() {
    const obsidianOptions = this.requestedCapabilities[OBSIDIAN_CAPABILITY_KEY];
    return obsidianOptions.appVersion;
  },
  /**
   * Returns the Obsidian installer version this test is running under.
   */
  getObsidianInstallerVersion() {
    const obsidianOptions = this.requestedCapabilities[OBSIDIAN_CAPABILITY_KEY];
    return obsidianOptions.installerVersion;
  },
  /**
   * Returns the ObsidianPage object with convenience helper functions.
   * You can also just import the page object directly with
   * ```ts
   * import { obsidianPage } from "wdio-obsidian-service"
   * ```
   */
  getObsidianPage() {
    return new ObsidianPage(this);
  }
};

// src/service.ts
import semver2 from "semver";
import _3 from "lodash";
var log2 = getLogger("wdio-obsidian-service");
function getDefaultCacheDir(rootDir) {
  return path4.resolve(rootDir, process.env.WEBDRIVER_CACHE_DIR ?? process.env.OBSIDIAN_CACHE ?? "./.obsidian-cache");
}
function getServiceErrorMessage(e) {
  return `Failed to download and setup Obsidian. Caused by:
${e.stack}
 ------The above causes:-----`;
}
function resolveEntry(rootDir, entry) {
  if (typeof entry == "string") {
    return path4.resolve(rootDir, entry);
  } else if ("path" in entry) {
    return { ...entry, path: path4.resolve(rootDir, entry.path) };
  } else {
    return entry;
  }
}
function selectPlugins(currentPlugins, selection) {
  if (selection !== void 0) {
    const unknownPlugins = _3.difference(selection, currentPlugins.map((p) => p.id));
    if (unknownPlugins.length > 0) {
      throw Error(`Unknown plugin ids: ${unknownPlugins.join(", ")}`);
    }
    return currentPlugins.map((p) => ({
      ...p,
      enabled: selection.includes(p.id) || p.id == "wdio-obsidian-service-plugin"
    }));
  } else {
    return currentPlugins;
  }
}
function selectThemes(currentThemes, selection) {
  if (selection !== void 0) {
    if (selection != "default" && currentThemes.every((t) => t.name != selection)) {
      throw Error(`Unknown theme: ${selection}`);
    }
    return currentThemes.map((t) => ({ ...t, enabled: selection != "default" && t.name === selection }));
  } else {
    return currentThemes;
  }
}
function getNormalizedObsidianOptions(cap) {
  return cap[OBSIDIAN_CAPABILITY_KEY];
}
var minSupportedObsidianVersion = "1.0.3";
var ObsidianLauncherService = class {
  constructor(options, capabilities, config) {
    this.options = options;
    this.capabilities = capabilities;
    this.config = config;
    this.rootDir = config.rootDir || process.cwd();
    this.obsidianLauncher = new ObsidianLauncher({
      cacheDir: config.cacheDir ?? getDefaultCacheDir(this.rootDir),
      versionsUrl: options.versionsUrl,
      communityPluginsUrl: options.communityPluginsUrl,
      communityThemesUrl: options.communityThemesUrl
    });
    this.helperPluginPath = path4.resolve(fileURLToPath3(import.meta.url), "../../helper-plugin");
  }
  /**
   * Validates wdio:obsidianOptions and downloads Obsidian, plugins, and themes.
   */
  async onPrepare(config, capabilities) {
    try {
      if (!Array.isArray(capabilities)) {
        capabilities = Object.values(capabilities).map(
          (multiremoteOption) => multiremoteOption.capabilities
        );
      }
      const obsidianCapabilities = capabilities.flatMap((cap) => {
        if ("browserName" in cap && cap.browserName === "obsidian") {
          return [cap];
        } else {
          return [];
        }
      });
      for (const cap of obsidianCapabilities) {
        const obsidianOptions = cap[OBSIDIAN_CAPABILITY_KEY] ?? {};
        const vault = obsidianOptions.vault ? path4.resolve(this.rootDir, obsidianOptions.vault) : void 0;
        if (vault && !fs2.existsSync(vault)) {
          throw Error(`Vault "${vault}" doesn't exist`);
        }
        const plugins = await this.obsidianLauncher.downloadPlugins(
          (obsidianOptions.plugins ?? []).concat([this.helperPluginPath]).map((p) => resolveEntry(this.rootDir, p))
        );
        const themes = await this.obsidianLauncher.downloadThemes(
          (obsidianOptions.themes ?? []).map((t) => resolveEntry(this.rootDir, t))
        );
        let appVersion = cap.browserVersion ?? cap[OBSIDIAN_CAPABILITY_KEY]?.appVersion ?? "latest";
        appVersion = (await this.obsidianLauncher.getVersionInfo(appVersion)).version;
        if (semver2.lt(appVersion, minSupportedObsidianVersion)) {
          throw Error(`Minimum supported Obsidian version is ${minSupportedObsidianVersion}`);
        }
        if (isAppium(cap)) {
          let apk = getAppiumOptions(cap).app;
          if (!apk) {
            apk = await this.obsidianLauncher.downloadAndroid(appVersion);
          }
          let chromedriverDir = getAppiumOptions(cap).chromedriverExecutableDir;
          if (!chromedriverDir) {
            chromedriverDir = path4.join(this.obsidianLauncher.cacheDir, "appium-chromedriver");
          }
          const normalizedObsidianOptions = {
            ...obsidianOptions,
            plugins,
            themes,
            vault,
            appVersion,
            installerVersion: appVersion,
            emulateMobile: false
          };
          cap[OBSIDIAN_CAPABILITY_KEY] = normalizedObsidianOptions;
          cap["appium:app"] = apk;
          cap["appium:chromedriverExecutableDir"] = chromedriverDir;
          cap["wdio:enforceWebDriverClassic"] = true;
          delete cap.browserName;
          delete cap.browserVersion;
          if (!getAppiumOptions(cap)["noReset"]) {
            console.warn("Note: For best performance set noReset to true, wdio-obsidian-service will handle resetting Obsidian between tests.");
          }
        } else {
          const [, installerVersion] = await this.obsidianLauncher.resolveVersion(
            appVersion,
            obsidianOptions.installerVersion ?? "earliest"
          );
          const installerInfo = await this.obsidianLauncher.getInstallerInfo(installerVersion);
          let installerPath;
          if (obsidianOptions.binaryPath) {
            installerPath = path4.resolve(this.rootDir, obsidianOptions.binaryPath);
          } else {
            installerPath = await this.obsidianLauncher.downloadInstaller(installerVersion);
          }
          let appPath;
          if (obsidianOptions.appPath) {
            appPath = path4.resolve(this.rootDir, obsidianOptions.appPath);
          } else {
            appPath = await this.obsidianLauncher.downloadApp(appVersion);
          }
          let chromedriverPath = cap["wdio:chromedriverOptions"]?.binary;
          if (!chromedriverPath) {
            chromedriverPath = await this.obsidianLauncher.downloadChromedriver(installerVersion);
          }
          const normalizedObsidianOptions = {
            ...obsidianOptions,
            plugins,
            themes,
            vault,
            binaryPath: installerPath,
            appPath,
            appVersion,
            installerVersion,
            emulateMobile: obsidianOptions.emulateMobile ?? false
          };
          cap.browserName = "chrome";
          cap.browserVersion = installerInfo.chrome;
          cap[OBSIDIAN_CAPABILITY_KEY] = normalizedObsidianOptions;
          cap["goog:chromeOptions"] = {
            binary: installerPath,
            windowTypes: ["app", "webview"],
            ...cap["goog:chromeOptions"],
            args: [
              // Workaround for SUID issue on linux. See https://github.com/electron/electron/issues/42510
              ...process.platform == "linux" ? ["--no-sandbox"] : [],
              ...cap["goog:chromeOptions"]?.args ?? []
            ]
          };
          cap["wdio:chromedriverOptions"] = {
            // allowedIps is not included in the types, but gets passed as --allowed-ips to chromedriver.
            // It defaults to ["0.0.0.0"] which makes Windows Firewall complain, and we don't need remote
            // connections anyways.
            allowedIps: [],
            ...cap["wdio:chromedriverOptions"],
            binary: chromedriverPath
          };
          cap["wdio:enforceWebDriverClassic"] = true;
        }
      }
    } catch (e) {
      throw new SevereServiceError(getServiceErrorMessage(e));
    }
  }
};
var ObsidianWorkerService = class {
  constructor(options, capabilities, config) {
    this.options = options;
    this.capabilities = capabilities;
    this.config = config;
    /** Path on Android devices to store temporary vaults */
    this.androidVaultDir = "/storage/emulated/0/Documents/wdio-obsidian-service-vaults";
    this.obsidianLauncher = new ObsidianLauncher({
      cacheDir: config.cacheDir ?? getDefaultCacheDir(config.rootDir || process.cwd()),
      versionsUrl: options.versionsUrl,
      communityPluginsUrl: options.communityPluginsUrl,
      communityThemesUrl: options.communityThemesUrl
    });
    this.tmpDirs = [];
  }
  /**
   * Creates a copy of the vault with plugins and themes installed
   */
  async setupVault(cap) {
    const obsidianOptions = getNormalizedObsidianOptions(cap);
    let vaultCopy;
    if (obsidianOptions.vault != void 0) {
      log2.info(`Opening vault ${obsidianOptions.vault}`);
      vaultCopy = await this.obsidianLauncher.setupVault({
        vault: obsidianOptions.vault,
        copy: true,
        plugins: obsidianOptions.plugins,
        themes: obsidianOptions.themes
      });
      this.tmpDirs.push(vaultCopy);
    } else {
      log2.info(`Opening Obsidian without a vault`);
    }
    obsidianOptions.vaultCopy = vaultCopy;
  }
  /**
   * Sets up the --user-data-dir for the Electron app. Sets the obsidian.json in the dir to open the vault on boot.
   */
  async electronSetupConfigDir(cap) {
    const obsidianOptions = getNormalizedObsidianOptions(cap);
    const configDir = await this.obsidianLauncher.setupConfigDir({
      appVersion: obsidianOptions.appVersion,
      installerVersion: obsidianOptions.installerVersion,
      appPath: obsidianOptions.appPath,
      vault: obsidianOptions.vaultCopy,
      // `app.emulateMobile` just sets this localStorage variable. Setting it ourselves here instead of calling
      // the function simplifies the boot/plugin load sequence and makes sure plugins load in mobile mode.
      localStorage: obsidianOptions.emulateMobile ? { "EmulateMobile": "1" } : {}
    });
    this.tmpDirs.push(configDir);
    cap["goog:chromeOptions"] = {
      ...cap["goog:chromeOptions"],
      args: [
        `--user-data-dir=${configDir}`,
        ...(cap["goog:chromeOptions"]?.args ?? []).filter((arg) => {
          const match = arg.match(/^--user-data-dir=(.*)$/);
          return !match || !this.tmpDirs.includes(match[1]);
        })
      ]
    };
  }
  /**
   * Sets up the Obsidian app. Installs and launches it if needed, and sets permissions and contexts.
   * 
   * You can configure Appium to install and launch the app for you. However, if you do that it reboots the app after
   * every session/spec which is really slow. So we are handling the app setup manually here.
   */
  async appiumSetupApp() {
    const browser2 = this.browser;
    const appiumOptions = getAppiumOptions(browser2.requestedCapabilities);
    const obsidianOptions = getNormalizedObsidianOptions(browser2.requestedCapabilities);
    const appId = "md.obsidian";
    const dumpsys = await browser2.execute("mobile: shell", { command: "dumpsys", args: ["package", appId] });
    const installedObsidianVersion = dumpsys.match(/versionName[=:](.*)$/m)?.[1]?.trim();
    if (installedObsidianVersion != obsidianOptions.appVersion) {
      await browser2.execute("mobile: terminateApp", { appId });
      await browser2.execute("mobile: removeApp", { appId, keepData: false });
      await browser2.execute("mobile: installApp", {
        appPath: appiumOptions.app,
        // the APK
        timeout: appiumOptions.androidInstallTimeout,
        // respect appium configuration
        grantPermissions: true
        // this and autoGrantPermissions don't seem to really work, see below
      });
    }
    const appState = await browser2.execute("mobile: queryAppState", { appId });
    if (appState < 4) {
      await browser2.execute("mobile: activateApp", { appId });
    }
    await browser2.execute("mobile: changePermissions", {
      action: "grant",
      appPackage: appId,
      permissions: "all"
    });
    await browser2.execute("mobile: changePermissions", {
      action: "allow",
      appPackage: appId,
      permissions: [
        // these are from apk AndroidManifest.xml (extracted with apktool)
        "READ_EXTERNAL_STORAGE",
        "WRITE_EXTERNAL_STORAGE",
        "MANAGE_EXTERNAL_STORAGE"
      ],
      target: "appops"
      // requires appium --allow-insecure adb_shell
    });
    const context = "WEBVIEW_md.obsidian";
    await browser2.waitUntil(
      async () => (await browser2.getContexts()).includes(context),
      { timeout: 3e4, interval: 100 }
    );
    await browser2.switchContext(context);
    await browser2.execute(() => {
      if (window.location.href != "http://localhost/") {
        window.location.replace("http://localhost/");
      }
    });
    if (!obsidianOptions.vault) {
      await this.appiumCloseVault();
    }
  }
  /**
   * Opens the vault in appium.
   */
  async appiumOpenVault() {
    const browser2 = this.browser;
    const obsidianOptions = getNormalizedObsidianOptions(browser2.requestedCapabilities);
    const androidVault = `${this.androidVaultDir}/${path4.basename(obsidianOptions.vaultCopy)}`;
    obsidianOptions.uploadedVault = androidVault;
    await appiumUploadFiles(browser2, { src: obsidianOptions.vaultCopy, dest: androidVault });
    await browser2.execute(async (androidVault2) => {
      localStorage.clear();
      localStorage.setItem("mobile-external-vaults", JSON.stringify([androidVault2]));
      localStorage.setItem("mobile-selected-vault", androidVault2);
      localStorage.setItem(`enable-plugin-${androidVault2}`, "true");
    }, androidVault);
    await navigateAndWait(browser2, () => location.reload());
  }
  /**
   * Close any open vault and go back to the vault switcher
   */
  async appiumCloseVault() {
    const browser2 = this.browser;
    if (await browser2.execute(() => localStorage.length > 0)) {
      await browser2.execute(() => localStorage.clear());
      await navigateAndWait(browser2, () => location.reload());
    }
  }
  /**
   * Waits for Obsidian to be ready, and does some other final setup.
   */
  async prepareApp() {
    const browser2 = this.browser;
    const obsidianOptions = getNormalizedObsidianOptions(browser2.requestedCapabilities);
    if (obsidianOptions.emulateMobile && obsidianOptions.vault != void 0) {
      await browser2.waitUntil(
        () => browser2.execute(() => !!window.electron),
        { timeout: 1e4, interval: 100 }
      );
      const [width, height] = await browser2.execute(() => [window.innerWidth, window.innerHeight]);
      await browser2.execute(async (width2, height2) => {
        await window.electron.remote.getCurrentWindow().setSize(width2, height2);
      }, width, height);
    }
    if (obsidianOptions.vault) {
      await browser2.waitUntil(
        // wait until the helper plugin is loaded
        () => browser2.execute(() => !!window.wdioObsidianService),
        { timeout: 6e4, interval: 100 }
      );
      await browser2.executeObsidian(async ({ app }) => {
        await new Promise((resolve2) => app.workspace.onLayoutReady(resolve2));
      });
    } else {
      await browser2.execute(async () => {
        if (document.readyState === "loading") {
          return new Promise((resolve2) => document.addEventListener("DOMContentLoaded", () => resolve2()));
        }
      });
    }
  }
  createReloadObsidian() {
    const service = this;
    const reloadObsidian = async function({ vault, plugins, theme } = {}) {
      const oldObsidianOptions = getNormalizedObsidianOptions(this.requestedCapabilities);
      const selectedPlugins = selectPlugins(oldObsidianOptions.plugins, plugins);
      const selectedThemes = selectThemes(oldObsidianOptions.themes, theme);
      if (!vault && oldObsidianOptions.vaultCopy == void 0) {
        throw Error(`No vault is open, pass a vault path to reloadObsidian`);
      }
      const newObsidianOptions = {
        ...oldObsidianOptions,
        // Resolve relative to PWD instead of root dir during tests
        vault: vault ? path4.resolve(vault) : oldObsidianOptions.vault,
        plugins: selectedPlugins,
        themes: selectedThemes
      };
      if (isAppium(this.requestedCapabilities)) {
        this.requestedCapabilities[OBSIDIAN_CAPABILITY_KEY] = newObsidianOptions;
        if (vault) {
          await service.setupVault(this.requestedCapabilities);
          await service.appiumOpenVault();
        } else {
          await navigateAndWait(this, () => location.replace("http://localhost/_capacitor_file_/not-a-file"));
          const local = path4.join(oldObsidianOptions.vaultCopy, ".obsidian");
          const localCommunityPlugins = path4.join(local, "community-plugins.json");
          const localAppearance = path4.join(local, "appearance.json");
          const remote2 = `${oldObsidianOptions.uploadedVault}/.obsidian`;
          const remoteCommunityPlugins = `${remote2}/community-plugins.json`;
          const remoteAppearance = `${remote2}/appearance.json`;
          await appiumDownloadFile(this, remoteCommunityPlugins, localCommunityPlugins).catch(() => {
          });
          await appiumDownloadFile(this, remoteAppearance, localAppearance).catch(() => {
          });
          await service.obsidianLauncher.setupVault({
            vault: oldObsidianOptions.vaultCopy,
            copy: false,
            plugins: selectedPlugins,
            themes: selectedThemes
          });
          let files = [localCommunityPlugins, localAppearance];
          files = (await Promise.all(files.map(async (f) => await fileExists(f) ? f : ""))).filter((f) => f);
          await appiumUploadFiles(this, { src: local, dest: remote2, files });
          await navigateAndWait(this, () => location.replace("http://localhost/"));
        }
      } else {
        const newCap = _3.cloneDeep(
          _3.omit(this.requestedCapabilities, ["browserName", "browserVersion"])
        );
        newCap[OBSIDIAN_CAPABILITY_KEY] = newObsidianOptions;
        if (vault) {
          await service.setupVault(newCap);
          await service.electronSetupConfigDir(newCap);
          await this.reloadSession(newCap);
        } else {
          await this.pause(2e3);
          await this.deleteSession({ shutdownDriver: false });
          await service.obsidianLauncher.setupVault({
            vault: oldObsidianOptions.vaultCopy,
            copy: false,
            plugins: selectedPlugins,
            themes: selectedThemes
          });
          await this.reloadSession(newCap);
        }
      }
      await service.prepareApp();
    };
    return reloadObsidian;
  }
  /**
   * Runs before the session and browser have started.
   */
  async beforeSession(config, cap) {
    try {
      if (!cap[OBSIDIAN_CAPABILITY_KEY]) return;
      if (cap[OBSIDIAN_CAPABILITY_KEY].vault != void 0) {
        await this.setupVault(cap);
      }
      if (!isAppium(cap)) {
        await this.electronSetupConfigDir(cap);
      }
    } catch (e) {
      throw new SevereServiceError(getServiceErrorMessage(e));
    }
  }
  /**
   * Runs after session and browser have started, but before tests.
   */
  async before(cap, specs, browser2) {
    this.browser = browser2;
    try {
      if (!cap[OBSIDIAN_CAPABILITY_KEY]) return;
      const newBrowserCommands = {
        ...browserCommands,
        reloadObsidian: this.createReloadObsidian()
      };
      for (const [name, cmd] of Object.entries(newBrowserCommands)) {
        browser2[name] = cmd;
      }
      if (isAppium(browser2.requestedCapabilities)) {
        await this.appiumSetupApp();
        if (cap[OBSIDIAN_CAPABILITY_KEY].vault) {
          await this.appiumOpenVault();
        }
      }
      await this.prepareApp();
    } catch (e) {
      throw new SevereServiceError(getServiceErrorMessage(e));
    }
  }
  /** Runs after tests are done, but before the session is shut down */
  async after(result, cap) {
    const browser2 = this.browser;
    if (!cap[OBSIDIAN_CAPABILITY_KEY]) return;
    if (isAppium(cap)) {
      await this.appiumCloseVault();
      await browser2.execute("mobile: shell", { command: "rm", args: ["-rf", this.androidVaultDir] });
    }
  }
  /**
   * Cleanup
   */
  async afterSession() {
    for (const tmpDir of this.tmpDirs) {
      await fsAsync3.rm(tmpDir, { recursive: true, force: true });
    }
  }
};

// src/standalone.ts
init_esm_shims();
import { remote } from "webdriverio";
async function startWdioSession(params, serviceOptions) {
  serviceOptions = serviceOptions ?? {};
  const capabilities = params.capabilities;
  const testRunnerOptions = {
    cacheDir: params.cacheDir
  };
  const launcherService = new ObsidianLauncherService(
    serviceOptions,
    [capabilities],
    testRunnerOptions
  );
  const workerService = new ObsidianWorkerService(serviceOptions, capabilities, testRunnerOptions);
  await launcherService.onPrepare(testRunnerOptions, [capabilities]);
  await workerService.beforeSession(testRunnerOptions, capabilities);
  const browser2 = await remote(params);
  await workerService.before(capabilities, [], browser2);
  return browser2;
}

// src/index.ts
var index_default = ObsidianWorkerService;
var launcher = ObsidianLauncherService;
async function obsidianBetaAvailable(opts = {}) {
  opts = typeof opts == "string" ? { cacheDir: opts } : opts;
  const launcher2 = new ObsidianLauncher2(opts);
  const versionInfo = await launcher2.getVersionInfo("latest-beta");
  return versionInfo.isBeta && await launcher2.isAvailable(versionInfo.version);
}
var resolveObsidianVersions = deprecate(async function(appVersion, installerVersion, cacheDir) {
  const launcher2 = new ObsidianLauncher2({ cacheDir });
  return await launcher2.resolveVersion(appVersion, installerVersion);
}, "resolveObsidianVersions is deprecated, use parseObsidianVersions instead");
async function parseObsidianVersions(versions, opts = {}) {
  const launcher2 = new ObsidianLauncher2({ cacheDir: opts.cacheDir });
  return launcher2.parseVersions(versions);
}
export {
  index_default as default,
  launcher,
  minSupportedObsidianVersion,
  obsidianBetaAvailable,
  obsidianPage_default as obsidianPage,
  parseObsidianVersions,
  resolveObsidianVersions,
  startWdioSession
};
