// obsidian-launcher/dist/chunk-4OAOPZAM.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; } function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } async function _asyncNullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return await rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; } async function _asyncOptionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = await fn(value); } else if (op === 'call' || op === 'optionalCall') { value = await fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }

var _chunkUMVJ7W6Kcjs = require('./chunk-UMVJ7W6K.cjs');

// src/utils.ts
var _promises = require('fs/promises'); var _promises2 = _interopRequireDefault(_promises);
var _fs = require('fs'); var _fs2 = _interopRequireDefault(_fs);
var _path = require('path'); var _path2 = _interopRequireDefault(_path);
var _os = require('os'); var _os2 = _interopRequireDefault(_os);
var _consola = require('consola');
var _promisepool = require('@supercharge/promise-pool');
var _lodash = require('lodash'); var _lodash2 = _interopRequireDefault(_lodash);
var consola = _consola.createConsola.call(void 0, {
  throttle: -1,
  // disable throttle
  formatOptions: {
    date: false
  }
});
async function fileExists(path6) {
  try {
    await _promises2.default.stat(path6);
    return true;
  } catch (e) {
    if (_optionalChain([e, 'optionalAccess', _6 => _6.code]) == "ENOENT") {
      return false;
    }
    throw e;
  }
}
async function makeTmpDir(prefix) {
  return _promises2.default.mkdtemp(_path2.default.join(_os2.default.tmpdir(), _nullishCoalesce(prefix, () => ( "tmp-"))));
}
var logged = /* @__PURE__ */ new Map();
function warnOnce(key, message) {
  const times = _nullishCoalesce(logged.get(key), () => ( 0));
  if (times <= 0) {
    consola.warn({ message });
  }
  logged.set(key, times + 1);
}
async function atomicCreate(dest, func, opts = {}) {
  const { replace = true, preserveTmpDir = false } = opts;
  dest = _path2.default.resolve(dest);
  const parentDir = _path2.default.dirname(dest);
  if (!replace && await fileExists(dest)) return;
  await _promises2.default.mkdir(parentDir, { recursive: true });
  const scratch = await _promises2.default.mkdtemp(_path2.default.join(parentDir, `.${_path2.default.basename(dest)}.tmp.`));
  try {
    let result = await _asyncNullishCoalesce(await func(scratch), async () => ( scratch));
    result = _path2.default.resolve(scratch, result);
    if (!result.startsWith(scratch)) {
      throw new Error(`Returned path ${result} not under scratch`);
    }
    if (replace) {
      if (await _asyncOptionalChain([(await _promises2.default.stat(dest).catch(() => null)), 'optionalAccess', async _7 => _7.isDirectory, 'call', async _8 => _8()])) {
        await _promises2.default.rename(dest, `${scratch}.old`);
      }
      await _promises2.default.rename(result, dest);
    } else {
      if (!await fileExists(dest)) {
        await _promises2.default.rename(result, dest).catch((e) => {
          if (_optionalChain([e, 'optionalAccess', _9 => _9.code]) != "ENOTEMPTY") throw e;
        });
      }
    }
    await _promises2.default.rm(scratch, { recursive: true, force: true });
    await _promises2.default.rm(`${scratch}.old`, { recursive: true, force: true });
  } catch (e) {
    if (!preserveTmpDir) {
      await _promises2.default.rm(scratch, { recursive: true, force: true });
    }
    throw e;
  }
}
async function linkOrCp(src, dest) {
  await _promises2.default.rm(dest, { recursive: true, force: true });
  try {
    await _promises2.default.link(src, dest);
  } catch (e2) {
    await _promises2.default.copyFile(src, dest);
  }
}
function pathIsUnder(parent, child) {
  const rel = _path2.default.relative(_path2.default.resolve(parent), _path2.default.resolve(child));
  return rel != "" && rel.split(_path2.default.sep)[0] != ".." && !_path2.default.isAbsolute(rel);
}
async function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
async function withTimeout(promise, timeout) {
  let timer;
  const result = Promise.race([
    promise,
    new Promise((resolve, reject) => timer = setTimeout(() => reject(Error("Promise timed out")), timeout))
  ]);
  return result.finally(() => clearTimeout(timer));
}
async function pool(size, items, func) {
  const { results } = await _promisepool.PromisePool.for(items).withConcurrency(size).handleError(async (error) => {
    throw error;
  }).useCorrespondingResults().process(func);
  return results;
}
async function maybe(promise) {
  return promise.then((r) => ({ success: true, result: r, error: void 0 })).catch((e) => ({ success: false, result: void 0, error: e }));
}
async function until(func, opts) {
  const { timeout, poll = 100 } = opts;
  let time = 0;
  let result;
  let error;
  while (!result && time < timeout) {
    try {
      result = await func();
      error = void 0;
    } catch (e) {
      error = e;
    }
    if (!result) {
      await sleep(poll);
    }
    time += poll;
  }
  if (!result) {
    throw new Error("Timed out waiting for condition" + (error ? `: ${error}` : ""));
  }
  return result;
}
async function retry(func, opts = {}) {
  const { retries = 4, backoff = 1e3, retryIf = () => true } = opts;
  let attempt = 0;
  let error;
  while (attempt < retries) {
    try {
      return await func(attempt);
    } catch (e) {
      error = e;
    }
    const delay = backoff * Math.random() + backoff * Math.pow(2, attempt);
    attempt += 1;
    if (!retryIf(error) || attempt >= retries) {
      throw error;
    }
    await sleep(delay);
  }
  throw error;
}
function watchFiles(files, func, options) {
  const debouncedFunc = _lodash2.default.debounce((curr, prev) => {
    if (curr.mtimeMs > prev.mtimeMs || curr.mtimeMs == 0 && prev.mtimeMs != 0) {
      func(curr, prev);
    }
  }, options.debounce);
  for (const file of files) {
    _fs2.default.watchFile(file, { interval: options.interval, persistent: options.persistent }, debouncedFunc);
  }
}
function normalizeObject(canonical, obj) {
  const rootCanonical = canonical, rootObj = obj;
  function helper(canonical2, obj2) {
    if (_lodash2.default.isPlainObject(canonical2)) {
      if (_lodash2.default.isPlainObject(obj2)) {
        obj2 = _lodash2.default.pick(obj2, Object.keys(canonical2));
        obj2 = _lodash2.default.mapValues(obj2, (v, k) => helper(canonical2[k], v));
        obj2 = _lodash2.default.omitBy(obj2, (v) => v === void 0);
        return obj2;
      } else {
        return obj2;
      }
    } else if (canonical2 === null) {
      return obj2;
    } else {
      throw Error(`Invalid canonical form ${JSON.stringify(rootCanonical)}`);
    }
  }
  return helper(rootCanonical, rootObj);
}

// src/launcher.ts


var _crypto = require('crypto'); var _crypto2 = _interopRequireDefault(_crypto);
var _extractzip = require('extract-zip'); var _extractzip2 = _interopRequireDefault(_extractzip);
var _get = require('@electron/get');
var _child_process = require('child_process'); var _child_process2 = _interopRequireDefault(_child_process);

var _semver = require('semver'); var _semver2 = _interopRequireDefault(_semver);
var _url = require('url');

var _dotenv = require('dotenv'); var _dotenv2 = _interopRequireDefault(_dotenv);

// src/types.ts
var obsidianVersionsSchemaVersion = "2.0.0";

// src/apis.ts


var _promises3 = require('stream/promises');
var _stream = require('stream');

var _readlinesync = require('readline-sync'); var _readlinesync2 = _interopRequireDefault(_readlinesync);


var _process = require('process');
function parseLinkHeader(linkHeader) {
  function parseLinkData(linkData) {
    return Object.fromEntries(
      linkData.split(";").flatMap((x) => {
        const partMatch = x.trim().match(/^([^=]+?)\s*=\s*"?([^"]+)"?$/);
        return partMatch ? [[partMatch[1], partMatch[2]]] : [];
      })
    );
  }
  const linkDatas = linkHeader.split(/,\s*(?=<)/).flatMap((link) => {
    const linkMatch = link.trim().match(/^<([^>]*)>(.*)$/);
    if (linkMatch) {
      return [{
        url: linkMatch[1],
        ...parseLinkData(linkMatch[2])
      }];
    } else {
      return [];
    }
  }).filter((l) => l.rel);
  return Object.fromEntries(linkDatas.map((l) => [l.rel, l]));
}
function createURL(url, base, params = {}) {
  const cleanParams = _lodash2.default.call(void 0, params).pickBy((x) => x !== void 0).mapValues((v) => String(v)).value();
  const urlObj = new URL(url, base);
  const searchParams = new URLSearchParams({ ...Object.fromEntries(urlObj.searchParams), ...cleanParams });
  if ([...searchParams].length > 0) {
    urlObj.search = "?" + searchParams;
  }
  return urlObj.toString();
}
async function fetchGitHubAPI(url, params = {}) {
  url = createURL(url, "https://api.github.com", params);
  const token = _process.env.GITHUB_TOKEN;
  const headers = token ? { Authorization: "Bearer " + token } : {};
  const response = await fetch(url, { headers });
  if (!response.ok) {
    throw new Error(`GitHub API error: ${await response.text()}`);
  }
  return response;
}
async function fetchGitHubAPIPaginated(url, params = {}) {
  const results = [];
  let next = createURL(url, "https://api.github.com", { per_page: 100, ...params });
  while (next) {
    const response = await fetchGitHubAPI(next);
    results.push(...await response.json());
    next = _optionalChain([parseLinkHeader, 'call', _10 => _10(_nullishCoalesce(response.headers.get("link"), () => ( ""))), 'access', _11 => _11.next, 'optionalAccess', _12 => _12.url]);
  }
  return results;
}
async function obsidianApiLogin(opts) {
  const { interactive = false, savePath } = opts;
  const cached = savePath ? _dotenv2.default.parse(await _promises2.default.readFile(savePath).catch(() => "")) : {};
  let email = _nullishCoalesce(_process.env.OBSIDIAN_EMAIL, () => ( cached.OBSIDIAN_EMAIL));
  let password = _nullishCoalesce(_process.env.OBSIDIAN_PASSWORD, () => ( cached.OBSIDIAN_PASSWORD));
  let promptedCredentials = false;
  const predownloadMessage = "pre-download the Obsidian beta with:\n    npx obsidian-launcher download app -v <version>";
  if (!email || !password) {
    if (interactive) {
      consola.log("Obsidian Insiders account is required to download Obsidian beta versions.");
      consola.log("You can set OBSIDIAN_EMAIL and OBSIDIAN_PASSWORD environment variables (.env file is supported) to avoid this prompt.");
      email = email || _readlinesync2.default.question("Obsidian email: ");
      password = password || _readlinesync2.default.question("Obsidian password: ", { hideEchoBack: true });
      promptedCredentials = true;
    } else {
      throw Error(
        "Obsidian Insiders account is required to download Obsidian beta versions. Set the OBSIDIAN_EMAIL and OBSIDIAN_PASSWORD environment variables (.env file is supported)" + (_process.env.CI ? "." : ` or ${predownloadMessage}`)
      );
    }
  }
  function parseSignin(r) {
    return { token: r.token ? "token" : void 0, error: _optionalChain([r, 'access', _13 => _13.error, 'optionalAccess', _14 => _14.toString, 'call', _15 => _15()]), license: r.license };
  }
  let needsMfa = false;
  let retries = 0;
  let signin = void 0;
  while (!_optionalChain([signin, 'optionalAccess', _16 => _16.token]) && retries < 3) {
    if (retries > 0 || _process.env.CI) {
      await sleep(2 * Math.random() + retries * retries * 3);
    }
    let mfa = "";
    if (needsMfa && interactive) {
      mfa = _readlinesync2.default.question("Obsidian 2FA: ");
    }
    const response = await fetch("https://api.obsidian.md/user/signin", {
      method: "post",
      headers: {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.183 Safari/537.36",
        "Origin": "app://obsidian.md",
        "Content-Type": "application/json"
      },
      body: JSON.stringify({ email, password, mfa })
    }).then((r) => r.json());
    signin = parseSignin(response);
    const error = _optionalChain([signin, 'access', _17 => _17.error, 'optionalAccess', _18 => _18.toLowerCase, 'call', _19 => _19()]);
    if (_optionalChain([error, 'optionalAccess', _20 => _20.includes, 'call', _21 => _21("2fa")]) && !needsMfa) {
      needsMfa = true;
      if (!interactive) {
        throw Error(
          "Can't login with 2FA in a non-interactive session. To download Obsidian beta versions disable 2FA on your account" + (_process.env.CI ? "." : ` or ${predownloadMessage}`)
        );
      }
    } else if (["please wait", "try again"].some((m) => _optionalChain([error, 'optionalAccess', _22 => _22.includes, 'call', _23 => _23(m)]))) {
      consola.warn(`Obsidian login failed: ${signin.error}`);
      consola.warn("Retrying obsidian login...");
      retries++;
    } else if (!signin.token) {
      throw Error(`Obsidian login failed: ${_nullishCoalesce(signin.error, () => ( "unknown error"))}`);
    }
  }
  if (!_optionalChain([signin, 'optionalAccess', _24 => _24.token])) {
    throw Error(`Obsidian login failed: ${_nullishCoalesce(_optionalChain([signin, 'optionalAccess', _25 => _25.error]), () => ( "unknown error"))}`);
  } else if (!_optionalChain([signin, 'optionalAccess', _26 => _26.license])) {
    throw Error("Obsidian Insiders account is required to download Obsidian beta versions");
  }
  if (savePath && promptedCredentials) {
    const save = _readlinesync2.default.question("Cache credentails to disk? [y/n]: ");
    if (["y", "yes"].includes(save.toLowerCase())) {
      await _promises2.default.writeFile(
        savePath,
        `OBSIDIAN_EMAIL='${email}'
OBSIDIAN_PASSWORD='${password}'
`
      );
      consola.log(`Saved Obsidian credentials to ${_path2.default.relative(process.cwd(), savePath)}`);
    }
  }
  return signin.token;
}
async function fetchObsidianApi(url, opts) {
  if (!opts.token) {
    throw Error("Obsidian credentials required to download Obsidian beta release");
  }
  url = createURL(url, "https://releases.obsidian.md");
  const response = await fetch(url, {
    headers: {
      "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.183 Safari/537.36",
      "Origin": "app://obsidian.md",
      "Authorization": "Bearer " + opts.token
    }
  });
  return response;
}
async function downloadResponse(func, dest, opts = {}) {
  await retry(async () => {
    const response = await func();
    if (!response.ok) {
      const error = Error(`${response.url} failed with ${response.status}`);
      error.status = response.status;
      throw error;
    }
    const fileStream = _fs2.default.createWriteStream(dest, { flags: "w" });
    const fetchStream = _stream.Readable.fromWeb(response.body);
    await _promises3.finished.call(void 0, fetchStream.pipe(fileStream));
  }, {
    retryIf: (e) => !e.status || ![401, 403, 404].includes(e.status),
    ...opts
  });
}

// src/chromeLocalStorage.ts

var _classiclevel = require('classic-level');
var ChromeLocalStorage = class {
  /** Pass the path to the user data dir for Chrome/Electron. If it doesn't exist it will be created. */
  constructor(userDataDir) {
    this.userDataDir = userDataDir;
    this.encodeKey = (domain, key) => `_${domain}\0${key}`;
    this.decodeKey = (key) => {
      const parts = key.slice(1).split("\0");
      return [parts[0], parts.slice(1).join("\0")];
    };
    this.encodeValue = (value) => `${value}`;
    this.decodeValue = (value) => value.slice(1);
    this.db = new (0, _classiclevel.ClassicLevel)(_path2.default.join(userDataDir, "Local Storage/leveldb/"));
  }
  /**
   * Get a value from localStorage
   * @param domain Domain the value is under, e.g. "https://example.com" or "app://obsidian.md"
   * @param key Key to retreive
   */
  async getItem(domain, key) {
    const value = await this.db.get(this.encodeKey(domain, key));
    return value === void 0 ? null : this.decodeValue(value);
  }
  /**
   * Set a value in localStorage
   * @param domain Domain the value is under, e.g. "https://example.com" or "app://obsidian.md"
   * @param key Key to set
   * @param value Value to set
   */
  async setItem(domain, key, value) {
    await this.db.put(this.encodeKey(domain, key), this.encodeValue(value));
  }
  /**
   * Removes a key from localStorage
   * @param domain Domain the values is under, e.g. "https://example.com" or "app://obsidian.md"
   * @param key key to remove.
   */
  async removeItem(domain, key) {
    await this.db.del(this.encodeKey(domain, key));
  }
  /** Get all items in localStorage as [domain, key, value] tuples */
  async getAllItems() {
    const result = [];
    for await (const pair of this.db.iterator()) {
      if (pair[0].startsWith("_")) {
        const [domain, key] = this.decodeKey(pair[0]);
        const value = this.decodeValue(pair[1]);
        result.push([domain, key, value]);
      }
    }
    return result;
  }
  /**
   * Write multiple values to localStorage in batch
   * @param domain Domain the values are under, e.g. "https://example.com" or "app://obsidian.md"
   * @param data key/value mapping to write
   */
  async setItems(domain, data) {
    await this.db.batch(
      Object.entries(data).map(([key, value]) => ({
        type: "put",
        key: this.encodeKey(domain, key),
        value: this.encodeValue(value)
      }))
    );
  }
  /**
   * Close the localStorage database.
   */
  async close() {
    await this.db.close();
  }
};

// src/launcherUtils.ts




var _util = require('util');




var _zlib = require('zlib'); var _zlib2 = _interopRequireDefault(_zlib);

var _chromeremoteinterface = require('chrome-remote-interface'); var _chromeremoteinterface2 = _interopRequireDefault(_chromeremoteinterface);
var execFile = _util.promisify.call(void 0, _child_process2.default.execFile);
function normalizeGitHubRepo(repo) {
  return _nullishCoalesce(_optionalChain([repo, 'access', _27 => _27.match, 'call', _28 => _28(/^(https?:\/\/)?(github.com\/)?(.*?)\/?$/), 'optionalAccess', _29 => _29[3]]), () => ( repo));
}
async function extractGz(archive, dest) {
  await _promises3.pipeline.call(void 0, _fs2.default.createReadStream(archive), _zlib2.default.createGunzip(), _fs2.default.createWriteStream(dest));
}
async function sevenZ(args, options) {
  const sevenZipScript = _path2.default.resolve(_url.fileURLToPath.call(void 0, _chunkUMVJ7W6Kcjs.importMetaUrl), "../7z.js");
  const proc = _child_process2.default.spawn(process.execPath, [sevenZipScript, ...args], {
    stdio: "pipe",
    ...options
  });
  let stdout = "", stderr = "";
  proc.stdout.on("data", (data) => stdout += data);
  proc.stderr.on("data", (data) => stderr += data);
  const procExit = new Promise((resolve) => proc.on("close", (code) => resolve(_nullishCoalesce(code, () => ( -1)))));
  const exitCode = await procExit;
  const result = { stdout, stderr };
  if (exitCode != 0) {
    throw Error(`"7z ${args.join(" ")}" failed with ${exitCode}:
${stdout}
${stderr}`);
  }
  return result;
}
async function extractObsidianAppImage(appImage, dest) {
  await atomicCreate(dest, async (scratch) => {
    await sevenZ(["x", "-o.", _path2.default.relative(scratch, appImage)], { cwd: scratch });
    return scratch;
  });
}
async function extractObsidianTar(tar, dest) {
  await atomicCreate(dest, async (scratch) => {
    await extractGz(tar, _path2.default.join(scratch, "inflated.tar"));
    await sevenZ(["x", "-o.", "inflated.tar"], { cwd: scratch });
    return (await _promises2.default.readdir(scratch)).find((p) => p.match("obsidian-"));
  });
}
async function extractObsidianExe(exe, arch, dest) {
  let subArchive;
  if (arch == "x64") {
    subArchive = `$PLUGINSDIR/app-64.7z`;
  } else if (arch == "ia32") {
    subArchive = `$PLUGINSDIR/app-32.7z`;
  } else if (arch == "arm64") {
    subArchive = `$PLUGINSDIR/app-arm64.7z`;
  } else {
    throw Error(`No Obsidian installer found for ${process.platform} ${process.arch}`);
  }
  await atomicCreate(dest, async (scratch) => {
    await sevenZ(["x", "-oinstaller", _path2.default.relative(scratch, exe), subArchive], { cwd: scratch });
    await sevenZ(["x", "-oobsidian", _path2.default.join("installer", subArchive)], { cwd: scratch });
    return "obsidian";
  });
}
async function extractObsidianDmg(dmg, dest) {
  dest = _path2.default.resolve(dest);
  await atomicCreate(dest, async (scratch) => {
    if (process.platform == "darwin") {
      const proc = await execFile("hdiutil", ["attach", "-nobrowse", "-readonly", dmg]);
      const volume = proc.stdout.match(/\/Volumes\/.*$/m)[0];
      const files = await _promises2.default.readdir(volume);
      let obsidianApp = files.includes("Obsidian.app") ? "Obsidian.app" : _path2.default.join(files[0], "Obsidian.app");
      obsidianApp = _path2.default.join(volume, obsidianApp);
      try {
        await _promises2.default.cp(obsidianApp, scratch, { recursive: true, verbatimSymlinks: true, preserveTimestamps: true });
      } finally {
        await execFile("hdiutil", ["detach", volume]);
      }
      await execFile("xattr", ["-cr", scratch]);
      return scratch;
    } else {
      await sevenZ(["x", "-o.", _path2.default.relative(scratch, dmg), "*/Obsidian.app", "Obsidian.app"], { cwd: scratch });
      const files = await _promises2.default.readdir(scratch);
      const obsidianApp = files.includes("Obsidian.app") ? "Obsidian.app" : _path2.default.join(files[0], "Obsidian.app");
      return _path2.default.join(scratch, obsidianApp);
    }
  });
}
async function getCdpSession(launcher, params) {
  const [appVersion, installerVersion] = await launcher.resolveVersion(
    _nullishCoalesce(params.appVersion, () => ( "latest")),
    _nullishCoalesce(params.installerVersion, () => ( "latest"))
  );
  const cleanup = [];
  const doCleanup = async () => {
    for (const func of [...cleanup].reverse()) {
      await func();
    }
  };
  const pluginDir = _path2.default.join(params.vault, ".obsidian", "plugins", "obsidian-launcher");
  await _promises2.default.mkdir(pluginDir, { recursive: true });
  await _promises2.default.writeFile(_path2.default.join(pluginDir, "manifest.json"), JSON.stringify({
    id: "obsidian-launcher",
    name: "Obsidian Launcher",
    version: "1.0.0",
    minAppVersion: "0.0.1",
    description: "",
    author: "obsidian-launcher",
    isDesktopOnly: false
  }));
  await _promises2.default.writeFile(_path2.default.join(pluginDir, "main.js"), `
        const obsidian = require('obsidian');
        class ObsidianLauncherPlugin extends obsidian.Plugin {
            async onload() { window.obsidianLauncher = {app: this.app, obsidian: obsidian}; };
        }
        module.exports = ObsidianLauncherPlugin;
    `);
  const communityPluginsPath = _path2.default.join(params.vault, ".obsidian", "community-plugins.json");
  let communityPlugins = ["obsidian-launcher"];
  if (await fileExists(communityPluginsPath)) {
    communityPlugins = [...JSON.parse(await _promises2.default.readFile(communityPluginsPath, "utf-8")), ...communityPlugins];
  }
  await _promises2.default.writeFile(communityPluginsPath, JSON.stringify(communityPlugins));
  try {
    const launchResult = await launcher.launch({
      ...params,
      // will choose a random available port
      args: [`--remote-debugging-port=0`, "--test-type=webdriver", ..._nullishCoalesce(params.args, () => ( []))]
    });
    if (params.copy) {
      cleanup.push(() => _promises2.default.rm(launchResult.vault, { recursive: true, force: true }));
    }
    const { proc } = launchResult;
    cleanup.push(() => _promises2.default.rm(launchResult.configDir, { recursive: true, force: true }));
    const procExit = new Promise((resolve) => proc.on("close", (code) => resolve(_nullishCoalesce(code, () => ( -1)))));
    cleanup.push(async () => {
      proc.kill("SIGTERM");
      const timeout = await maybe(withTimeout(procExit, 5 * 1e3));
      if (!timeout.success) {
        consola.warn(`Stuck process ${proc.pid}, using SIGKILL`);
        proc.kill("SIGKILL");
      }
      await procExit;
    });
    const portPromise = new Promise((resolve, reject) => {
      void procExit.then(() => reject(Error("Processed ended without opening a port")));
      proc.stderr.on("data", (data) => {
        const port2 = _optionalChain([data, 'access', _30 => _30.toString, 'call', _31 => _31(), 'access', _32 => _32.match, 'call', _33 => _33(/ws:\/\/[\w.]+?:(\d+)/), 'optionalAccess', _34 => _34[1]]);
        if (port2) {
          resolve(Number(port2));
        }
      });
    });
    const port = await maybe(withTimeout(portPromise, 20 * 1e3));
    if (!port.success) {
      throw new Error("Timed out waiting for Chrome DevTools protocol port");
    }
    const client = await _chromeremoteinterface2.default.call(void 0, { port: port.result });
    cleanup.push(() => client.close());
    const expr = _semver2.default.gte(appVersion, "0.12.8") ? "!!window.obsidianLauncher" : "!!window.app.workspace";
    await until(
      () => client.Runtime.evaluate({ expression: expr }).then((r) => r.result.value),
      { timeout: 5e3 }
    );
    return {
      ...launchResult,
      client,
      cleanup: doCleanup,
      vault: launchResult.vault
    };
  } catch (e) {
    await doCleanup();
    throw e;
  }
}
async function cdpEvaluate(client, expression) {
  const response = await client.Runtime.evaluate({ expression, returnByValue: true });
  if (response.exceptionDetails) {
    throw Error(response.exceptionDetails.text);
  }
  return response.result.value;
}
async function cdpEvaluateUntil(client, expression, opts) {
  return await until(() => cdpEvaluate(client, expression), opts);
}
async function getProcesses() {
  if (process.platform === "win32") {
    const { stdout } = await execFile("powershell.exe", [
      "-NoProfile",
      "-ExecutionPolicy",
      "Bypass",
      "-Command",
      "Get-CimInstance Win32_Process | Sort-Object -Property CreationDate | Select-Object ProcessId,CommandLine | ConvertTo-Json"
    ]);
    const data = JSON.parse(stdout);
    const list = Array.isArray(data) ? data : [data];
    return list.map((p) => ({ pid: p.ProcessId, command: p.CommandLine || "" }));
  } else {
    const { stdout } = await execFile("ps", ["-xww", "-o", "lstart=,pid=,command="]);
    const processes = stdout.split("\n").map((l) => l.trim()).filter((line) => line).map((line) => {
      const [_5, startTime, pid, command] = line.match(/^(\w+\s+\w+\s+\d+\s+\d\d:\d\d:\d\d\s+\d\d\d\d)\s+(\d+)\s+(.*)$/);
      return {
        pid: Number(pid),
        startTime: new Date(startTime).getTime(),
        command
      };
    });
    return _lodash2.default.sortBy(processes, (i) => i.startTime).map((i) => _lodash2.default.omit(i, "startTime"));
  }
}
async function getObsidianCli(args) {
  const clean = (s) => {
    return s.trim().match(/^["']?(.*?)["']?$/)[1];
  };
  let processes = await getProcesses();
  processes = processes.filter((p) => p.command.includes("--tag=obsidian-launcher"));
  let obsidianInstances = [];
  for (const proc of processes) {
    try {
      const match2 = proc.command.match(/(.*?) --user-data-dir=(.*?obsidian-launcher-config-.+?)( |$)/);
      const [_5, exe, configDir] = match2.map(clean);
      const obsidianJson = JSON.parse(await _promises2.default.readFile(_path2.default.join(configDir, "obsidian.json"), "utf-8"));
      for (const [vaultId, vaultInfo] of Object.entries(obsidianJson.vaults)) {
        const vaultPath = await _promises2.default.realpath(vaultInfo.path);
        obsidianInstances.push({ pid: proc.pid, exe, configDir, vaultId, vaultPath });
      }
    } catch (e) {
      console.warn(`Failed to connect to obsidian-launcher instance ${proc.pid}: ${e}`);
    }
  }
  obsidianInstances = obsidianInstances.reverse();
  let match;
  let newArgs = [...args];
  if (args.length > 0 && args[0].startsWith("vault=")) {
    const vault = args[0].slice(6);
    match = obsidianInstances.find((i) => i.vaultId == vault || _path2.default.basename(i.vaultPath).toUpperCase() == vault.toUpperCase());
    if (!match) {
      const systemTmpDir = await _promises2.default.realpath(_os2.default.tmpdir());
      match = obsidianInstances.find((i) => pathIsUnder(systemTmpDir, i.vaultPath) && _optionalChain([_path2.default, 'access', _35 => _35.basename, 'call', _36 => _36(i.vaultPath), 'access', _37 => _37.toUpperCase, 'call', _38 => _38(), 'access', _39 => _39.match, 'call', _40 => _40(/(.*)-.{6}/), 'optionalAccess', _41 => _41[1]]) == vault.toUpperCase());
    }
    if (!match) {
      throw Error(`No running Obsidian instance for ${vault}`);
    }
    newArgs = args.slice(1);
  }
  if (!match) {
    const cwd = await _promises2.default.realpath(process.cwd()).catch(() => process.cwd());
    match = obsidianInstances.find((i) => cwd == i.vaultPath || pathIsUnder(i.vaultPath, cwd));
  }
  if (!match) {
    match = obsidianInstances.at(0);
  }
  if (!match) {
    throw Error(`No running Obsidian instance`);
  }
  newArgs = [
    `--user-data-dir=${match.configDir}`,
    ...process.platform == "linux" ? ["--no-sandbox"] : [],
    `vault=${match.vaultId}`,
    ...newArgs
  ];
  return [match.exe, newArgs];
}
async function fetchObsidianDesktopReleases(sinceDate, sinceSha) {
  const repo = "obsidianmd/obsidian-releases";
  let commitHistory = await fetchGitHubAPIPaginated(`repos/${repo}/commits`, {
    path: "desktop-releases.json",
    since: sinceDate
  });
  commitHistory.reverse();
  if (sinceSha) {
    commitHistory = _lodash2.default.takeRightWhile(commitHistory, (c) => c.sha != sinceSha);
  }
  const fileHistory = await pool(
    4,
    commitHistory,
    (commit) => fetch(`https://raw.githubusercontent.com/${repo}/${commit.sha}/desktop-releases.json`).then((r) => r.json())
  );
  const commitDate = _nullishCoalesce(_optionalChain([commitHistory, 'access', _42 => _42.at, 'call', _43 => _43(-1), 'optionalAccess', _44 => _44.commit, 'access', _45 => _45.committer, 'access', _46 => _46.date]), () => ( sinceDate));
  const commitSha = _nullishCoalesce(_optionalChain([commitHistory, 'access', _47 => _47.at, 'call', _48 => _48(-1), 'optionalAccess', _49 => _49.sha]), () => ( sinceSha));
  return [fileHistory, { commitDate, commitSha }];
}
async function fetchObsidianGitHubReleases() {
  const releases = await fetchGitHubAPIPaginated(`repos/obsidianmd/obsidian-releases/releases`);
  return releases.reverse();
}
var BROKEN_VERSIONS = [
  "0.12.16",
  // broken download link
  "1.4.7",
  // broken download link
  "1.4.8",
  // broken download link
  "1.0.1"
  // won't launch
];
function parseObsidianDesktopRelease(fileRelease) {
  const parse = (r, isBeta) => {
    return {
      version: r.latestVersion,
      isBeta,
      downloads: {
        asar: r.downloadUrl
      }
    };
  };
  const result = { current: parse(fileRelease, false) };
  if (fileRelease.beta && fileRelease.beta.latestVersion !== fileRelease.latestVersion) {
    result.beta = parse(fileRelease.beta, true);
  }
  return result;
}
function parseObsidianGithubRelease(gitHubRelease) {
  const version = gitHubRelease.name;
  const assets = gitHubRelease.assets.map((a) => ({
    url: a.browser_download_url,
    digest: _nullishCoalesce(a.digest, () => ( `id:${a.id}`))
  }));
  const asar = assets.find((a) => a.url.match(`${version}.asar.gz$`));
  const appImage = assets.find((a) => a.url.match(`${version}.AppImage$`));
  const appImageArm = assets.find((a) => a.url.match(`${version}-arm64.AppImage$`));
  const tar = assets.find((a) => a.url.match(`${version}.tar.gz$`));
  const tarArm = assets.find((a) => a.url.match(`${version}-arm64.tar.gz$`));
  const dmg = assets.find((a) => a.url.match(`${version}(-universal)?.dmg$`));
  const exe = assets.find((a) => a.url.match(`${version}.exe$`));
  const apk = assets.find((a) => a.url.match(`${version}.apk$`));
  return {
    version,
    gitHubRelease: gitHubRelease.html_url,
    downloads: {
      asar: _optionalChain([asar, 'optionalAccess', _50 => _50.url]),
      appImage: _optionalChain([appImage, 'optionalAccess', _51 => _51.url]),
      appImageArm: _optionalChain([appImageArm, 'optionalAccess', _52 => _52.url]),
      tar: _optionalChain([tar, 'optionalAccess', _53 => _53.url]),
      tarArm: _optionalChain([tarArm, 'optionalAccess', _54 => _54.url]),
      dmg: _optionalChain([dmg, 'optionalAccess', _55 => _55.url]),
      exe: _optionalChain([exe, 'optionalAccess', _56 => _56.url]),
      apk: _optionalChain([apk, 'optionalAccess', _57 => _57.url])
    },
    installers: {
      appImage: appImage ? { digest: appImage.digest } : void 0,
      appImageArm: appImageArm ? { digest: appImageArm.digest } : void 0,
      tar: tar ? { digest: tar.digest } : void 0,
      tarArm: tarArm ? { digest: tarArm.digest } : void 0,
      dmg: dmg ? { digest: dmg.digest } : void 0,
      exe: exe ? { digest: exe.digest } : void 0
    }
  };
}
var INSTALLER_KEYS = [
  "appImage",
  "appImageArm",
  "tar",
  "tarArm",
  "dmg",
  "exe"
];
async function extractInstallerInfo(version, installerKey, url) {
  const installerName = url.split("/").at(-1);
  consola.log(`Extrating installer info for ${installerName}...`);
  const tmpDir = await makeTmpDir("obsidian-launcher-");
  try {
    const installerPath = _path2.default.join(tmpDir, url.split("/").at(-1));
    await downloadResponse(() => fetch(url), installerPath);
    const exractedPath = _path2.default.join(tmpDir, "Obsidian");
    let platforms = [];
    if (installerKey == "appImage" || installerKey == "appImageArm") {
      await extractObsidianAppImage(installerPath, exractedPath);
      platforms = ["linux-" + (installerKey == "appImage" ? "x64" : "arm64")];
    } else if (installerKey == "tar" || installerKey == "tarArm") {
      await extractObsidianTar(installerPath, exractedPath);
      platforms = ["linux-" + (installerKey == "tar" ? "x64" : "arm64")];
    } else if (installerKey == "exe") {
      await extractObsidianExe(installerPath, "x64", exractedPath);
      const { stdout } = await sevenZ(["l", "-ba", _path2.default.relative(tmpDir, installerPath)], { cwd: tmpDir });
      const lines = stdout.trim().split("\n").map((l) => l.trim());
      const files = lines.map((l) => l.split(/\s+/).at(-1).replace(/\\/g, "/"));
      if (files.includes("$PLUGINSDIR/app-arm64.7z")) platforms.push("win32-arm64");
      if (files.includes("$PLUGINSDIR/app-32.7z")) platforms.push("win32-ia32");
      if (files.includes("$PLUGINSDIR/app-64.7z")) platforms.push("win32-x64");
    } else if (installerKey == "dmg") {
      await extractObsidianDmg(installerPath, exractedPath);
      platforms = ["darwin-arm64", "darwin-x64"];
    } else {
      throw new Error(`Unknown installer key ${installerKey}`);
    }
    const matches = [];
    const installerFiles = await _promises2.default.readdir(exractedPath, { recursive: true, withFileTypes: true });
    for (const file of installerFiles) {
      if (file.isFile() && !file.name.endsWith(".asar")) {
        const stream = _fs2.default.createReadStream(_path2.default.join(file.parentPath, file.name), { encoding: "utf-8" });
        let prev = "";
        for await (let chunk of stream) {
          const regex = /Chrome\/\d+\.\d+\.\d+\.\d+|Electron\/\d+\.\d+\.\d+/g;
          chunk = prev + chunk;
          matches.push(...[...(prev + chunk).matchAll(regex)].map((m) => m[0]));
          prev = chunk.slice(-64);
        }
      }
    }
    const versionSortKey = (v) => v.split(".").map((s) => s.padStart(9, "0")).join(".");
    const versions = _lodash2.default.call(void 0, matches).map((m) => m.split("/")).groupBy(0).mapValues((ms) => ms.map((m) => m[1])).mapValues((ms) => _lodash2.default.sortBy(ms, versionSortKey).at(-1)).value();
    const electron = versions["Electron"];
    const chrome = versions["Chrome"];
    if (!electron || !chrome) {
      throw new Error(`Failed to extract Electron and Chrome versions from binary ${installerPath}`);
    }
    consola.log(`Extracted installer info for ${installerName}`);
    return { electron, chrome, platforms };
  } finally {
    await _promises2.default.rm(tmpDir, { recursive: true, force: true });
  }
}
async function checkCompatibility(launcher, appVersion, installerVersion) {
  [appVersion, installerVersion] = await launcher.resolveVersion(appVersion, installerVersion);
  consola.log(`Checking if app ${appVersion} and installer ${installerVersion} are compatible...`);
  await launcher.downloadApp(appVersion);
  await launcher.downloadInstaller(installerVersion);
  const vault = await makeTmpDir("obsidian-launcher-");
  const cdpResult = await maybe(retry(
    () => getCdpSession(launcher, { appVersion, installerVersion, vault }),
    { retries: 3, backoff: 4e3 }
  ));
  if (!cdpResult.success) {
    consola.log(`app ${appVersion} with installer ${installerVersion} failed to launch: ${cdpResult.error}`);
    return false;
  }
  const { client, cleanup } = cdpResult.result;
  let result = true;
  try {
    if (_semver2.default.lt(appVersion, "0.7.4")) {
      result = _semver2.default.gte(installerVersion, "0.6.4");
    } else if (_semver2.default.lt(appVersion, "0.13.4")) {
      await cdpEvaluate(client, `window.app.commands.executeCommandById('app:open-settings')`);
      await until(() => cdpEvaluate(client, `
                document.evaluate(
                    "//*[contains(@class, 'vertical-tab-nav-item') and contains(text(),'About')]",
                    document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null
                ).singleNodeValue.click() ?? true;
            `), { timeout: 5e3 });
      const aboutText = await until(() => cdpEvaluate(client, `
                document.evaluate(
                    "//*[contains(@class, 'setting-item-name') and contains(text(),'Current version:')]",
                    document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null
                ).singleNodeValue.parentNode.innerText;
            `), { timeout: 5e3 });
      if (["manual installation", "manually install"].some((t) => aboutText.includes(t))) {
        result = false;
      }
    } else {
      await cdpEvaluate(client, `window.obsidianLauncher.app.commands.executeCommandById('app:show-debug-info')`);
      const debugInfo = await cdpEvaluateUntil(
        client,
        `document.querySelector(".debug-textarea").value.trim()`,
        { timeout: 5e3 }
      );
      if (debugInfo.toLowerCase().match(/installer version too low/)) {
        result = false;
      }
    }
  } finally {
    await cleanup();
    await _promises2.default.rm(vault, { recursive: true, force: true });
  }
  consola.log(`app ${appVersion} and installer ${installerVersion} are ${!result ? "in" : ""}compatible`);
  return result;
}
async function getCompatibilityInfos(versions, { _checkCompatibility = checkCompatibility } = {}) {
  const tmp = await makeTmpDir("obsidian-installer-compat-");
  try {
    const versionsFile = {
      metadata: {
        schemaVersion: obsidianVersionsSchemaVersion,
        commitDate: "1970-01-01T00:00:00Z",
        commitSha: "0000000000000000000000000000000000000000",
        timestamp: "1970-01-01T00:00:00Z"
      },
      versions: versions.map((v) => ({
        ...v,
        minInstallerVersion: _nullishCoalesce(v.minInstallerVersion, () => ( "0.0.0")),
        maxInstallerVersion: _nullishCoalesce(v.maxInstallerVersion, () => ( "999.9.9"))
      }))
    };
    await _promises2.default.writeFile(_path2.default.join(tmp, "obsidian-versions.json"), JSON.stringify(versionsFile));
    const launcher = new ObsidianLauncher({
      cacheDir: _path2.default.join(tmp, "cache"),
      versionsUrl: _url.pathToFileURL.call(void 0, _path2.default.join(tmp, "obsidian-versions.json")).toString()
    });
    const versionArr = _lodash2.default.call(void 0, _lodash2.default.cloneDeep(versions)).sort((a, b) => _semver2.default.compare(a.version, b.version)).dropWhile((v) => !v.downloads.appImage).filter((v) => !!v.downloads.asar).value();
    let maxInstallerVersion = void 0;
    for (const version of versionArr) {
      if (version.downloads.appImage) {
        maxInstallerVersion = version.version;
      }
      version.maxInstallerVersion = maxInstallerVersion;
    }
    const installerArr = versionArr.filter((v) => !!v.downloads.appImage);
    const installerIndexMap = _lodash2.default.fromPairs(installerArr.map((v, i) => [v.version, i]));
    for (const [i, version] of versionArr.entries()) {
      if (version.minInstallerVersion) {
        continue;
      }
      const prev = i > 0 ? versionArr[i - 1] : void 0;
      let start = prev ? installerIndexMap[prev.minInstallerVersion] : 0;
      let end = installerIndexMap[version.maxInstallerVersion];
      while (start <= end) {
        const mid = Math.floor((start + end) / 2);
        const compatible = await _checkCompatibility(
          launcher,
          version.version,
          installerArr[mid].version
        );
        if (!compatible) {
          start = mid + 1;
        } else {
          end = mid - 1;
        }
      }
      if (start > installerIndexMap[version.maxInstallerVersion]) {
        throw Error(`${version.version} failed to launch for all installers`);
      }
      version.minInstallerVersion = installerArr[start].version;
    }
    const origVersions = _lodash2.default.call(void 0, versions).map((v) => _lodash2.default.pick(v, ["version", "minInstallerVersion", "maxInstallerVersion"])).keyBy((v) => v.version).value();
    return versionArr.map((v) => ({
      version: v.version,
      minInstallerVersion: v.minInstallerVersion,
      maxInstallerVersion: v.maxInstallerVersion
    })).filter((v) => !_lodash2.default.isEqual(v, origVersions[v.version]));
  } finally {
    await _promises2.default.rm(tmp, { recursive: true, force: true });
  }
}
function normalizeObsidianVersionInfo(versionInfo) {
  versionInfo = _lodash2.default.cloneDeep(versionInfo);
  versionInfo.electronVersion = _optionalChain([versionInfo, 'access', _58 => _58.installers, 'optionalAccess', _59 => _59.appImage, 'optionalAccess', _60 => _60.electron]);
  versionInfo.chromeVersion = _optionalChain([versionInfo, 'access', _61 => _61.installers, 'optionalAccess', _62 => _62.appImage, 'optionalAccess', _63 => _63.chrome]);
  versionInfo.downloads = _nullishCoalesce(versionInfo.downloads, () => ( {}));
  versionInfo.installers = _nullishCoalesce(versionInfo.installers, () => ( {}));
  const canonicalForm = {
    version: null,
    minInstallerVersion: null,
    maxInstallerVersion: null,
    isBeta: null,
    gitHubRelease: null,
    downloads: {
      asar: null,
      appImage: null,
      appImageArm: null,
      tar: null,
      tarArm: null,
      dmg: null,
      exe: null,
      apk: null
    },
    installers: {
      appImage: { digest: null, electron: null, chrome: null, platforms: null },
      appImageArm: { digest: null, electron: null, chrome: null, platforms: null },
      tar: { digest: null, electron: null, chrome: null, platforms: null },
      tarArm: { digest: null, electron: null, chrome: null, platforms: null },
      dmg: { digest: null, electron: null, chrome: null, platforms: null },
      exe: { digest: null, electron: null, chrome: null, platforms: null }
    },
    electronVersion: null,
    chromeVersion: null
  };
  return normalizeObject(canonicalForm, versionInfo);
}
async function updateObsidianVersionList(original, {
  maxInstances = 1,
  _fetchObsidianDesktopReleases = fetchObsidianDesktopReleases,
  _fetchObsidianGitHubReleases = fetchObsidianGitHubReleases,
  _extractInstallerInfo = extractInstallerInfo,
  _checkCompatibility = checkCompatibility
} = {}) {
  const oldVersions = _lodash2.default.keyBy(_nullishCoalesce(_optionalChain([original, 'optionalAccess', _64 => _64.versions]), () => ( [])), (v) => v.version);
  let newVersions = _lodash2.default.cloneDeep(oldVersions);
  const [destkopReleases, commitInfo] = await _fetchObsidianDesktopReleases(
    _optionalChain([original, 'optionalAccess', _65 => _65.metadata, 'access', _66 => _66.commitDate]),
    _optionalChain([original, 'optionalAccess', _67 => _67.metadata, 'access', _68 => _68.commitSha])
  );
  for (const destkopRelease of destkopReleases) {
    const { current, beta } = parseObsidianDesktopRelease(destkopRelease);
    if (beta) {
      newVersions[beta.version] = _lodash2.default.merge(_nullishCoalesce(newVersions[beta.version], () => ( {})), beta);
    }
    newVersions[current.version] = _lodash2.default.merge(_nullishCoalesce(newVersions[current.version], () => ( {})), current);
  }
  const gitHubReleases = await _fetchObsidianGitHubReleases();
  for (const githubRelease of gitHubReleases) {
    if (_semver2.default.valid(githubRelease.name) && !_semver2.default.prerelease(githubRelease.name)) {
      const parsed = parseObsidianGithubRelease(githubRelease);
      const newVersion = _lodash2.default.merge(_nullishCoalesce(newVersions[parsed.version], () => ( {})), parsed);
      for (const installerKey of INSTALLER_KEYS) {
        const oldDigest = _optionalChain([oldVersions, 'access', _69 => _69[parsed.version], 'optionalAccess', _70 => _70.installers, 'access', _71 => _71[installerKey], 'optionalAccess', _72 => _72.digest]);
        const newDigest = _optionalChain([newVersion, 'access', _73 => _73.installers, 'optionalAccess', _74 => _74[installerKey], 'optionalAccess', _75 => _75.digest]);
        if (oldDigest && oldDigest != newDigest) {
          newVersion.installers[installerKey] = { digest: newDigest };
        }
      }
      newVersions[parsed.version] = newVersion;
    }
  }
  newVersions = _lodash2.default.omitBy(newVersions, (v) => BROKEN_VERSIONS.includes(v.version));
  const newInstallers = Object.values(newVersions).flatMap((v) => INSTALLER_KEYS.map((k) => [v, k])).filter(([v, key]) => _optionalChain([v, 'access', _76 => _76.downloads, 'optionalAccess', _77 => _77[key]]) && !_optionalChain([v, 'access', _78 => _78.installers, 'optionalAccess', _79 => _79[key], 'optionalAccess', _80 => _80.chrome]));
  const installerInfos = await pool(maxInstances, newInstallers, async ([v, key]) => {
    const installerInfo = await _extractInstallerInfo(v.version, key, v.downloads[key]);
    return { version: v.version, installers: { [key]: installerInfo } };
  });
  for (const installerInfo of installerInfos) {
    newVersions[installerInfo.version] = _lodash2.default.merge(_nullishCoalesce(newVersions[installerInfo.version], () => ( {})), installerInfo);
  }
  const compatInfos = await getCompatibilityInfos(
    Object.values(newVersions),
    { _checkCompatibility }
  );
  for (const compatInfo of compatInfos) {
    newVersions[compatInfo.version] = _lodash2.default.merge(_nullishCoalesce(newVersions[compatInfo.version], () => ( {})), compatInfo);
  }
  const result = {
    metadata: {
      schemaVersion: obsidianVersionsSchemaVersion,
      commitDate: commitInfo.commitDate,
      commitSha: commitInfo.commitSha,
      timestamp: _nullishCoalesce(_optionalChain([original, 'optionalAccess', _81 => _81.metadata, 'access', _82 => _82.timestamp]), () => ( ""))
      // set down below
    },
    versions: Object.values(newVersions).map(normalizeObsidianVersionInfo).sort((a, b) => _semver2.default.compare(a.version, b.version))
  };
  const dayMs = 24 * 60 * 60 * 1e3;
  const timeSinceLastUpdate = (/* @__PURE__ */ new Date()).getTime() - new Date(_nullishCoalesce(_optionalChain([original, 'optionalAccess', _83 => _83.metadata, 'access', _84 => _84.timestamp]), () => ( 0))).getTime();
  if (!_lodash2.default.isEqual(original, result) || timeSinceLastUpdate > 29 * dayMs) {
    result.metadata.timestamp = (/* @__PURE__ */ new Date()).toISOString();
  }
  return result;
}

// src/launcher.ts
var currentPlatform = {
  platform: process.platform,
  arch: process.arch
};
_dotenv2.default.config({ path: [".env"], quiet: true });
var minSupportedObsidianVersion = "0.12.8";
var ObsidianLauncher = class {
  /**
   * Construct an ObsidianLauncher.
   * @param opts.cacheDir Path to the cache directory. Defaults to "OBSIDIAN_CACHE" env var or ~/.obsidian-cache.
   * @param opts.versionsUrl Custom `obsidian-versions.json` url. Can be a file URL.
   * @param opts.communityPluginsUrl Custom `community-plugins.json` url. Can be a file URL.
   * @param opts.communityThemesUrl Custom `community-css-themes.json` url. Can be a file URL.
   * @param opts.cacheDuration If the cached version list is older than this (in ms), refetch it. Defaults to 30 minutes.
   * @param opts.interactive If it can prompt the user for input (e.g. for Obsidian credentials). Default false.
   */
  constructor(opts = {}) {
    this.interactive = false;
    this.cacheDir = _path2.default.resolve(
      _nullishCoalesce(_nullishCoalesce(opts.cacheDir, () => ( process.env.OBSIDIAN_CACHE)), () => ( _path2.default.join(_os2.default.homedir(), ".obsidian-cache")))
    );
    const defaultVersionsUrl = "https://raw.githubusercontent.com/jesse-r-s-hines/wdio-obsidian-service/HEAD/obsidian-versions.json";
    this.versionsUrl = _nullishCoalesce(opts.versionsUrl, () => ( defaultVersionsUrl));
    const defaultCommunityPluginsUrl = "https://raw.githubusercontent.com/obsidianmd/obsidian-releases/HEAD/community-plugins.json";
    this.communityPluginsUrl = _nullishCoalesce(opts.communityPluginsUrl, () => ( defaultCommunityPluginsUrl));
    const defaultCommunityThemesUrl = "https://raw.githubusercontent.com/obsidianmd/obsidian-releases/HEAD/community-css-themes.json";
    this.communityThemesUrl = _nullishCoalesce(opts.communityThemesUrl, () => ( defaultCommunityThemesUrl));
    this.cacheDuration = _nullishCoalesce(opts.cacheDuration, () => ( 30 * 60 * 1e3));
    this.interactive = _nullishCoalesce(opts.interactive, () => ( false));
    this.metadataCache = {};
  }
  /**
   * Returns file content fetched from url as JSON. Caches content to dest and uses that cache if its more recent than
   * cacheDuration ms or if there are network errors.
   */
  async cachedFetch(url, dest, cacheValid) {
    cacheValid = _nullishCoalesce(cacheValid, () => ( (() => true)));
    dest = _path2.default.join(this.cacheDir, dest);
    if (!(dest in this.metadataCache)) {
      let data;
      let error;
      const cacheMtime = await _asyncOptionalChain([(await _promises2.default.stat(dest).catch(() => void 0)), 'optionalAccess', async _85 => _85.mtime]);
      if (url.startsWith("file:")) {
        data = JSON.parse(await _promises2.default.readFile(_url.fileURLToPath.call(void 0, url), "utf-8"));
      }
      if (!data && cacheMtime && (/* @__PURE__ */ new Date()).getTime() - cacheMtime.getTime() < this.cacheDuration) {
        const parsed = JSON.parse(await _promises2.default.readFile(dest, "utf-8"));
        if (cacheValid(parsed)) {
          data = parsed;
        }
      }
      if (!data) {
        const response = await maybe(fetch(url).then(async (r) => {
          if (!r.ok) throw Error(`Fetch ${url} failed with status ${r.status}`);
          const d = await r.text();
          if (_lodash2.default.isError(_lodash2.default.attempt(JSON.parse, d))) throw Error(`Failed to parse response from ${url}`);
          return d;
        }));
        if (response.success) {
          await atomicCreate(dest, async (scratch) => {
            await _promises2.default.writeFile(_path2.default.join(scratch, "download.json"), response.result);
            return _path2.default.join(scratch, "download.json");
          }, { replace: true });
          data = JSON.parse(response.result);
        } else {
          error = response.error;
        }
      }
      if (!data && await fileExists(dest)) {
        const parsed = JSON.parse(await _promises2.default.readFile(dest, "utf-8"));
        if (cacheValid(parsed)) {
          consola.warn(error);
          consola.warn(`Unable to download ${url}, using cached file.`);
          data = parsed;
        }
      }
      if (!data) {
        throw error;
      }
      this.metadataCache[dest] = data;
    }
    return this.metadataCache[dest];
  }
  /**
   * Get parsed content of the current project's manifest.json
   */
  async getRootManifest() {
    if (!("manifest.json" in this.metadataCache)) {
      const root = _path2.default.parse(process.cwd()).root;
      let dir = process.cwd();
      while (dir != root && !await fileExists(_path2.default.join(dir, "manifest.json"))) {
        dir = _path2.default.dirname(dir);
      }
      const manifestPath = _path2.default.join(dir, "manifest.json");
      if (await fileExists(manifestPath)) {
        this.metadataCache["manifest.json"] = JSON.parse(await _promises2.default.readFile(manifestPath, "utf-8"));
      } else {
        this.metadataCache["manifest.json"] = null;
      }
    }
    return this.metadataCache["manifest.json"];
  }
  /**
   * Get information about all available Obsidian versions.
   */
  async getVersions() {
    const isValid = (d) => _semver2.default.satisfies(_nullishCoalesce(d.metadata.schemaVersion, () => ( "1.0.0")), `^${obsidianVersionsSchemaVersion}`);
    const versions = await this.cachedFetch(this.versionsUrl, "obsidian-versions.json", isValid);
    if (!isValid(versions)) {
      throw new Error(`${this.versionsUrl} format has changed, please update obsidian-launcher and wdio-obsidian-service`);
    }
    return versions.versions;
  }
  /**
   * Get information about all available community plugins.
   */
  async getCommunityPlugins() {
    return await this.cachedFetch(this.communityPluginsUrl, "obsidian-community-plugins.json");
  }
  /**
   * Get information about all available community themes.
   */
  async getCommunityThemes() {
    return await this.cachedFetch(this.communityThemesUrl, "obsidian-community-css-themes.json");
  }
  /**
   * Resolves Obsidian app and installer version strings to absolute versions.
   * @param appVersion specific version or one of
   *   - "latest": Get the current latest non-beta Obsidian version
   *   - "latest-beta": Get the current latest beta Obsidian version (or latest if there is no current beta)
   *   - "earliest": Get the `minAppVersion` set in your `manifest.json`
   * @param installerVersion specific version or one of
   *   - "latest": Get the latest Obsidian installer compatible with `appVersion`
   *   - "earliest": Get the oldest Obsidian installer compatible with `appVersion`
   * 
   * See also: [Obsidian App vs Installer Versions](../README.md#obsidian-app-vs-installer-versions)
   *
   * @returns [appVersion, installerVersion] with any "latest" etc. resolved to specific versions.
   */
  async resolveVersion(appVersion, installerVersion = "latest") {
    const versions = await this.getVersions();
    const appVersionInfo = await this.getVersionInfo(appVersion);
    appVersion = appVersionInfo.version;
    let installerVersionInfo;
    const { platform, arch } = process;
    if (_semver2.default.lt(appVersion, minSupportedObsidianVersion)) {
      warnOnce(
        "unsupported-version",
        `Obsidian versions before ${minSupportedObsidianVersion} are unsupported, some obsidian-launcher features will not work.`
      );
    }
    if (installerVersion == "latest") {
      installerVersionInfo = _lodash2.default.findLast(
        versions,
        (v) => _semver2.default.lte(v.version, appVersionInfo.version) && !!this.getInstallerKey(v, { platform, arch })
      );
    } else if (installerVersion == "earliest") {
      installerVersionInfo = versions.find(
        (v) => appVersionInfo.minInstallerVersion && _semver2.default.gte(v.version, appVersionInfo.minInstallerVersion) && _semver2.default.lte(v.version, appVersionInfo.version) && !!this.getInstallerKey(v, { platform, arch })
      );
    } else {
      installerVersion = _nullishCoalesce(_semver2.default.valid(installerVersion), () => ( installerVersion));
      installerVersionInfo = versions.find((v) => v.version == installerVersion);
    }
    if (!installerVersionInfo) {
      if (["earliest", "latest"].includes(installerVersion)) {
        throw Error(`No compatible installers available for Obsidian ${appVersion} for ${platform}-${arch}`);
      } else {
        throw Error(`No Obsidian installer ${installerVersion} found`);
      }
    }
    if (!appVersionInfo.minInstallerVersion || !appVersionInfo.maxInstallerVersion || _semver2.default.lt(installerVersionInfo.version, appVersionInfo.minInstallerVersion) || _semver2.default.gt(installerVersionInfo.version, appVersionInfo.maxInstallerVersion)) {
      warnOnce(
        `incompatible-versions-${appVersionInfo.version}-${installerVersionInfo.version}`,
        `App and installer versions are incompatible: app ${appVersionInfo.version} is compatible with installer >=${appVersionInfo.minInstallerVersion} <=${appVersionInfo.maxInstallerVersion} but ${installerVersionInfo.version} specified.`
      );
    }
    return [appVersionInfo.version, installerVersionInfo.version];
  }
  /**
   * Gets details about an Obsidian version.
   * @param appVersion Obsidian app version
   */
  async getVersionInfo(appVersion) {
    const versions = await this.getVersions();
    if (appVersion == "latest-beta") {
      appVersion = versions.at(-1).version;
    } else if (appVersion == "latest") {
      appVersion = versions.filter((v) => !v.isBeta).at(-1).version;
    } else if (appVersion == "earliest") {
      const manifest = await this.getRootManifest();
      if (!_optionalChain([manifest, 'optionalAccess', _86 => _86.minAppVersion])) {
        throw Error('Unable to resolve Obsidian appVersion "earliest", no manifest.json or minAppVersion found.');
      }
      appVersion = manifest.minAppVersion;
    } else {
      appVersion = _nullishCoalesce(_semver2.default.valid(appVersion), () => ( appVersion));
    }
    const versionInfo = versions.find((v) => v.version == appVersion);
    if (!versionInfo) {
      throw Error(`No Obsidian app version "${appVersion}" found`);
    }
    return versionInfo;
  }
  /**
   * Parses a string of Obsidian versions into [appVersion, installerVersion] tuples.
   * 
   * `versions` should be a space separated list of Obsidian app versions. You can optionally specify the installer
   * version by using "appVersion/installerVersion" e.g. `"1.7.7/1.8.10"`.
   * 
   * Example: 
   * ```js
   * launcher.parseVersions("1.8.10/1.7.7 latest latest-beta/earliest")
   * ```
   * 
   * See also: [Obsidian App vs Installer Versions](../README.md#obsidian-app-vs-installer-versions)
   * 
   * @param versions string to parse
   * @returns [appVersion, installerVersion][] resolved to specific versions.
   */
  async parseVersions(versions) {
    const parsedVersions = versions.split(/[ ,]/).filter((v) => v).map((v) => {
      const [appVersion, installerVersion = "earliest"] = v.split("/");
      return [appVersion, installerVersion];
    });
    const resolvedVersions = [];
    for (const [appVersion, installerVersion] of parsedVersions) {
      resolvedVersions.push(await this.resolveVersion(appVersion, installerVersion));
    }
    return _lodash2.default.uniqBy(resolvedVersions, (v) => v.join("/"));
  }
  getInstallerKey(installerVersionInfo, opts = {}) {
    const { platform, arch } = _lodash2.default.defaults({}, opts, currentPlatform);
    const platformName = `${platform}-${arch}`;
    const key = _lodash2.default.findKey(installerVersionInfo.installers, (v) => v && v.platforms.includes(platformName));
    return key;
  }
  /**
   * Gets details about the Obsidian installer for the given platform.
   * @param installerVersion Obsidian installer version
   * @param opts.platform Platform/os (defaults to host platform)
   * @param opts.arch Architecture (defaults to host architecture)
   */
  async getInstallerInfo(installerVersion, opts = {}) {
    const { platform, arch } = _lodash2.default.defaults({}, opts, currentPlatform);
    const versionInfo = await this.getVersionInfo(installerVersion);
    const key = this.getInstallerKey(versionInfo, { platform, arch });
    if (key) {
      return { ...versionInfo.installers[key], url: versionInfo.downloads[key] };
    } else {
      throw Error(
        `No Obsidian installer for ${installerVersion} ${platform}-${arch}` + (versionInfo.isBeta ? ` (${installerVersion} is a beta version)` : "")
      );
    }
  }
  /**
   * Log into the Obsidian api using your Insider's account so you can download beta versions.
   *
   * login will be called automatically when using downloadApp on an Obsidian beta version so you usually won't need
   * to call this directly.
   */
  async login() {
    if (!this.obsidianApiToken) {
      this.obsidianApiToken = await obsidianApiLogin({
        interactive: this.interactive,
        savePath: _path2.default.join(this.cacheDir, "obsidian-credentials.env")
      });
    }
  }
  /**
   * Downloads the Obsidian installer for the given version and platform/arch (defaults to host platform/arch).
   * Returns the file path.
   * @param installerVersion Obsidian installer version to download
   * @param opts.platform Platform/os of the installer to download (defaults to host platform)
   * @param opts.arch Architecture of the installer to download (defaults to host architecture)
   */
  async downloadInstaller(installerVersion, opts = {}) {
    const { platform, arch } = _lodash2.default.defaults({}, opts, currentPlatform);
    const versionInfo = await this.getVersionInfo(installerVersion);
    installerVersion = versionInfo.version;
    const installerInfo = await this.getInstallerInfo(installerVersion, { platform, arch });
    const installerDir = _path2.default.join(this.cacheDir, `obsidian-installer/${platform}-${arch}/Obsidian-${installerVersion}`);
    let binaryPath;
    let extractor;
    if (platform == "linux") {
      binaryPath = _path2.default.join(installerDir, "obsidian");
      extractor = (installer, dest) => extractObsidianAppImage(installer, dest);
    } else if (platform == "win32") {
      binaryPath = _path2.default.join(installerDir, "Obsidian.exe");
      extractor = (installer, dest) => extractObsidianExe(installer, arch, dest);
    } else if (platform == "darwin") {
      binaryPath = _path2.default.join(installerDir, "Contents/MacOS/Obsidian");
      extractor = (installer, dest) => extractObsidianDmg(installer, dest);
    } else {
      throw Error(`Unsupported platform ${platform}`);
    }
    await atomicCreate(installerDir, async (scratch) => {
      consola.log(`Downloading Obsidian installer v${installerVersion}...`);
      const installer = _path2.default.join(scratch, "installer");
      await downloadResponse(() => fetch(installerInfo.url), installer);
      const extracted = _path2.default.join(scratch, "extracted");
      await extractor(installer, extracted);
      return extracted;
    }, { replace: false });
    return binaryPath;
  }
  /**
   * Downloads the Obsidian asar for the given version. Returns the file path.
   * 
   * To download Obsidian beta versions you'll need to have an Obsidian Insiders account and either set the 
   * `OBSIDIAN_EMAIL` and `OBSIDIAN_PASSWORD` environment variables (`.env` file is supported) or pre-download the
   * Obsidian beta with `npx obsidian-launcher download app -v latest-beta`
   * 
   * @param appVersion Obsidian version to download
   */
  async downloadApp(appVersion) {
    const versionInfo = await this.getVersionInfo(appVersion);
    const appUrl = versionInfo.downloads.asar;
    if (!appUrl) {
      throw Error(`No asar found for Obsidian version ${appVersion}`);
    }
    const appPath = _path2.default.join(this.cacheDir, "obsidian-app", `obsidian-${versionInfo.version}.asar`);
    const isInsiders = new URL(appUrl).hostname.endsWith(".obsidian.md");
    if (isInsiders && !await fileExists(appPath)) {
      await this.login();
    }
    await atomicCreate(appPath, async (scratch) => {
      consola.log(`Downloading Obsidian app v${versionInfo.version} ...`);
      const archive = _path2.default.join(scratch, "app.asar.gz");
      const asar = _path2.default.join(scratch, "app.asar");
      if (isInsiders) {
        await downloadResponse(() => fetchObsidianApi(appUrl, { token: this.obsidianApiToken }), archive);
      } else {
        await downloadResponse(() => fetch(appUrl), archive);
      }
      await extractGz(archive, asar);
      return asar;
    }, { replace: false });
    return appPath;
  }
  /**
   * Downloads chromedriver for the given Obsidian version.
   * 
   * wdio will download chromedriver from the Chrome for Testing API automatically (see
   * https://github.com/GoogleChromeLabs/chrome-for-testing#json-api-endpoints). However, Google has only put
   * chromedriver since v115.0.5763.0 in that API, so wdio can't automatically download older versions of chromedriver
   * for old Electron versions. Here we download chromedriver for older versions ourselves using the @electron/get
   * package which fetches chromedriver from https://github.com/electron/electron/releases.
   * 
   * @param installerVersion Obsidian installer version
   */
  async downloadChromedriver(installerVersion, opts = {}) {
    const { platform, arch } = _lodash2.default.defaults({}, opts, currentPlatform);
    const installerInfo = await this.getInstallerInfo(installerVersion, { platform, arch });
    const chromedriverDir = _path2.default.join(this.cacheDir, `electron-chromedriver/${platform}-${arch}/${installerInfo.electron}`);
    let chromedriverPath;
    if (process.platform == "win32") {
      chromedriverPath = _path2.default.join(chromedriverDir, `chromedriver.exe`);
    } else {
      chromedriverPath = _path2.default.join(chromedriverDir, `chromedriver`);
    }
    await atomicCreate(chromedriverDir, async (scratch) => {
      consola.log(`Downloading chromedriver for electron ${installerInfo.electron} ...`);
      const chromedriverZipPath = await _get.downloadArtifact.call(void 0, {
        version: installerInfo.electron,
        artifactName: "chromedriver",
        cacheRoot: _path2.default.join(scratch, "download")
      });
      const extracted = _path2.default.join(scratch, "extracted");
      await _extractzip2.default.call(void 0, chromedriverZipPath, { dir: extracted });
      return extracted;
    }, { replace: false });
    return chromedriverPath;
  }
  /**
   * Downloads the Obsidian apk.
   */
  async downloadAndroid(version) {
    const versionInfo = await this.getVersionInfo(version);
    const apkUrl = versionInfo.downloads.apk;
    if (!apkUrl) {
      throw Error(
        `No apk found for Obsidian version ${version}` + (versionInfo.isBeta ? ` (${version} is a beta version)` : "")
      );
    }
    const apkPath = _path2.default.join(this.cacheDir, "obsidian-apk", `obsidian-${versionInfo.version}.apk`);
    await atomicCreate(apkPath, async (scratch) => {
      consola.log(`Downloading Obsidian apk v${versionInfo.version} ...`);
      const dest = _path2.default.join(scratch, "obsidian.apk");
      await downloadResponse(() => fetch(apkUrl), dest);
      return dest;
    }, { replace: false });
    return apkPath;
  }
  /** Gets the latest version of a plugin. */
  async getLatestPluginVersion(repo) {
    repo = normalizeGitHubRepo(repo);
    const manifestUrl = `https://raw.githubusercontent.com/${repo}/HEAD/manifest.json`;
    const cacheDest = _path2.default.join("obsidian-plugins", repo, "latest.json");
    const manifest = await this.cachedFetch(manifestUrl, cacheDest);
    return manifest.version;
  }
  /**
   * Downloads a plugin from a GitHub repo to the cache.
   * @param repo Repo
   * @param version Version of the plugin to install or "latest"
   * @returns path to the downloaded plugin
   */
  async downloadGitHubPlugin(repo, version = "latest") {
    repo = normalizeGitHubRepo(repo);
    if (version == "latest") {
      version = await this.getLatestPluginVersion(repo);
    }
    if (!_semver2.default.valid(version)) {
      throw Error(`Invalid version "${version}"`);
    }
    version = _semver2.default.valid(version);
    const pluginDir = _path2.default.join(this.cacheDir, "obsidian-plugins", repo, version);
    await atomicCreate(pluginDir, async (scratch) => {
      const assetsToDownload = { "manifest.json": true, "main.js": true, "styles.css": false };
      await Promise.all(
        Object.entries(assetsToDownload).map(async ([file, required]) => {
          const url = `https://github.com/${repo}/releases/download/${version}/${file}`;
          try {
            await downloadResponse(() => fetch(url), _path2.default.join(scratch, file));
          } catch (e3) {
            if (required) {
              throw Error(`No ${file} found for ${repo} version ${version}`);
            }
          }
        })
      );
      return scratch;
    }, { replace: false });
    return pluginDir;
  }
  /**
   * Downloads a community plugin to the cache.
   * @param id Id of the plugin
   * @param version Version of the plugin to install, or "latest"
   * @returns path to the downloaded plugin
   */
  async downloadCommunityPlugin(id, version = "latest") {
    const communityPlugins = await this.getCommunityPlugins();
    const pluginInfo = communityPlugins.find((p) => p.id == id);
    if (!pluginInfo) {
      throw Error(`No plugin with id ${id} found.`);
    }
    return await this.downloadGitHubPlugin(pluginInfo.repo, version);
  }
  /**
   * Downloads a list of plugins to the cache and returns a list of {@link DownloadedPluginEntry} with the downloaded
   * paths. Also adds the `id` property to the plugins based on the manifest.
   * 
   * You can download plugins from GitHub using `{repo: "org/repo"}` and community plugins using `{id: 'plugin-id'}`.
   * Local plugins will just be passed through.
   * 
   * @param plugins List of plugins to download.
   */
  async downloadPlugins(plugins) {
    return await Promise.all(
      plugins.map(async (plugin) => {
        if (typeof plugin == "object" && "originalType" in plugin) {
          return { ...plugin };
        }
        let pluginPath;
        let originalType;
        if (typeof plugin == "string") {
          pluginPath = _path2.default.resolve(plugin);
          originalType = "local";
        } else if ("path" in plugin) {
          ;
          pluginPath = _path2.default.resolve(plugin.path);
          originalType = "local";
        } else if ("repo" in plugin) {
          pluginPath = await this.downloadGitHubPlugin(plugin.repo, plugin.version);
          originalType = "github";
        } else if ("id" in plugin) {
          pluginPath = await this.downloadCommunityPlugin(plugin.id, plugin.version);
          originalType = "community";
        } else {
          throw Error("You must specify one of plugin path, repo, or id");
        }
        const manifestPath = _path2.default.join(pluginPath, "manifest.json");
        if (!await fileExists(manifestPath)) {
          throw Error(`No plugin found at ${pluginPath}`);
        }
        let pluginId = typeof plugin == "object" && "id" in plugin ? plugin.id : void 0;
        if (!pluginId) {
          pluginId = JSON.parse(await _promises2.default.readFile(manifestPath, "utf8").catch(() => "{}")).id;
          if (!pluginId) {
            throw Error(`${manifestPath} malformed.`);
          }
        }
        if (!await fileExists(_path2.default.join(pluginPath, "main.js"))) {
          throw Error(`No main.js found under ${pluginPath}`);
        }
        let enabled;
        if (typeof plugin == "string") {
          enabled = true;
        } else {
          enabled = _nullishCoalesce(plugin.enabled, () => ( true));
        }
        return { path: pluginPath, id: pluginId, enabled, originalType };
      })
    );
  }
  /**
   * Installs plugins into an Obsidian vault
   * @param vault Path to the vault to install the plugins in
   * @param plugins List plugins to install
   */
  async installPlugins(vault, plugins) {
    const downloadedPlugins = await this.downloadPlugins(plugins);
    const obsidianDir = _path2.default.join(vault, ".obsidian");
    await _promises2.default.mkdir(obsidianDir, { recursive: true });
    const enabledPluginsPath = _path2.default.join(obsidianDir, "community-plugins.json");
    let originalEnabledPlugins = [];
    if (await fileExists(enabledPluginsPath)) {
      originalEnabledPlugins = JSON.parse(await _promises2.default.readFile(enabledPluginsPath, "utf-8"));
    }
    let enabledPlugins = [...originalEnabledPlugins];
    for (const { path: pluginPath, enabled = true, originalType } of downloadedPlugins) {
      const manifestPath = _path2.default.join(pluginPath, "manifest.json");
      const pluginId = JSON.parse(await _promises2.default.readFile(manifestPath, "utf8").catch(() => "{}")).id;
      if (!pluginId) {
        throw Error(`${manifestPath} missing or malformed.`);
      }
      const pluginDest = _path2.default.join(obsidianDir, "plugins", pluginId);
      await _promises2.default.mkdir(pluginDest, { recursive: true });
      const files = {
        "manifest.json": true,
        "main.js": true,
        "styles.css": false
      };
      for (const [file, required] of Object.entries(files)) {
        if (await fileExists(_path2.default.join(pluginPath, file))) {
          await linkOrCp(_path2.default.join(pluginPath, file), _path2.default.join(pluginDest, file));
        } else if (required) {
          throw Error(`${pluginPath}/${file} missing.`);
        } else {
          await _promises2.default.rm(_path2.default.join(pluginDest, file), { force: true });
        }
      }
      if (await fileExists(_path2.default.join(pluginPath, "data.json"))) {
        await _promises2.default.cp(_path2.default.join(pluginPath, "data.json"), _path2.default.join(pluginDest, "data.json"));
      }
      const pluginAlreadyListed = enabledPlugins.includes(pluginId);
      if (enabled && !pluginAlreadyListed) {
        enabledPlugins.push(pluginId);
      } else if (!enabled && pluginAlreadyListed) {
        enabledPlugins = enabledPlugins.filter((p) => p != pluginId);
      }
      if (originalType == "local") {
        await _promises2.default.writeFile(_path2.default.join(pluginDest, ".hotreload"), "");
      }
    }
    if (!_lodash2.default.isEqual(enabledPlugins, originalEnabledPlugins)) {
      await _promises2.default.writeFile(enabledPluginsPath, JSON.stringify(enabledPlugins, void 0, 2));
    }
  }
  /** Gets the latest version of a theme. */
  async getLatestThemeVersion(repo) {
    repo = normalizeGitHubRepo(repo);
    const manifestUrl = `https://raw.githubusercontent.com/${repo}/HEAD/manifest.json`;
    const cacheDest = _path2.default.join("obsidian-themes", repo, "latest.json");
    const manifest = await this.cachedFetch(manifestUrl, cacheDest);
    return manifest.version;
  }
  /**
   * Downloads a theme from a GitHub repo to the cache.
   * @param repo Repo
   * @returns path to the downloaded theme
   */
  async downloadGitHubTheme(repo, version = "latest") {
    repo = normalizeGitHubRepo(repo);
    const latest = await this.getLatestThemeVersion(repo);
    if (version == "latest") {
      version = latest;
    }
    if (!_semver2.default.valid(version)) {
      throw Error(`Invalid version "${version}"`);
    }
    version = _semver2.default.valid(version);
    const themeDir = _path2.default.join(this.cacheDir, "obsidian-themes", repo, version);
    await atomicCreate(themeDir, async (scratch) => {
      const assetsToDownload = ["manifest.json", "theme.css"];
      let baseUrl = `https://github.com/${repo}/releases/download/${version}`;
      if (!(await fetch(`${baseUrl}/manifest.json`)).ok) {
        if (version != latest) {
          throw Error(`No theme version "${version}" found`);
        }
        baseUrl = `https://raw.githubusercontent.com/${repo}/HEAD`;
      }
      await Promise.all(
        assetsToDownload.map(
          async (file) => {
            const url = `${baseUrl}/${file}`;
            try {
              await downloadResponse(() => fetch(url), _path2.default.join(scratch, file));
            } catch (e4) {
              throw Error(`No ${file} found for ${repo}`);
            }
          }
        )
      );
    }, { replace: false });
    return themeDir;
  }
  /**
   * Downloads a community theme to the cache.
   * @param name name of the theme
   * @returns path to the downloaded theme
   */
  async downloadCommunityTheme(name, version = "latest") {
    const communityThemes = await this.getCommunityThemes();
    const themeInfo = communityThemes.find((p) => p.name == name);
    if (!themeInfo) {
      throw Error(`No theme with name ${name} found.`);
    }
    return await this.downloadGitHubTheme(themeInfo.repo, version);
  }
  /**
   * Downloads a list of themes to the cache and returns a list of {@link DownloadedThemeEntry} with the downloaded
   * paths. Also adds the `name` property to the plugins based on the manifest.
   * 
   * You can download themes from GitHub using `{repo: "org/repo"}` and community themes using `{name: 'theme-name'}`.
   * Local themes will just be passed through.
   * 
   * @param themes List of themes to download
   */
  async downloadThemes(themes) {
    return await Promise.all(
      themes.map(async (theme) => {
        if (typeof theme == "object" && "originalType" in theme) {
          return { ...theme };
        }
        let themePath;
        let originalType;
        if (typeof theme == "string") {
          themePath = _path2.default.resolve(theme);
          originalType = "local";
        } else if ("path" in theme) {
          ;
          themePath = _path2.default.resolve(theme.path);
          originalType = "local";
        } else if ("repo" in theme) {
          themePath = await this.downloadGitHubTheme(theme.repo, theme.version);
          originalType = "github";
        } else if ("name" in theme) {
          themePath = await this.downloadCommunityTheme(theme.name, theme.version);
          originalType = "community";
        } else {
          throw Error("You must specify one of theme path, repo, or name");
        }
        const manifestPath = _path2.default.join(themePath, "manifest.json");
        if (!await fileExists(manifestPath)) {
          throw Error(`No theme found at ${themePath}`);
        }
        let themeName = typeof theme == "object" && "name" in theme ? theme.name : void 0;
        if (!themeName) {
          const manifestPath2 = _path2.default.join(themePath, "manifest.json");
          themeName = JSON.parse(await _promises2.default.readFile(manifestPath2, "utf8").catch(() => "{}")).name;
          if (!themeName) {
            throw Error(`${themePath}/manifest.json malformed.`);
          }
        }
        if (!await fileExists(_path2.default.join(themePath, "theme.css"))) {
          throw Error(`No theme.css found under ${themePath}`);
        }
        let enabled;
        if (typeof theme == "string") {
          enabled = true;
        } else {
          enabled = _nullishCoalesce(theme.enabled, () => ( true));
        }
        return { path: themePath, name: themeName, enabled, originalType };
      })
    );
  }
  /**
   * Installs themes into an Obsidian vault
   * @param vault Path to the theme to install the themes in
   * @param themes List of themes to install
   */
  async installThemes(vault, themes) {
    const downloadedThemes = await this.downloadThemes(themes);
    const obsidianDir = _path2.default.join(vault, ".obsidian");
    await _promises2.default.mkdir(obsidianDir, { recursive: true });
    let enabledTheme = void 0;
    for (const { path: themePath, enabled = true } of downloadedThemes) {
      const manifestPath = _path2.default.join(themePath, "manifest.json");
      const cssPath = _path2.default.join(themePath, "theme.css");
      const themeName = JSON.parse(await _promises2.default.readFile(manifestPath, "utf8").catch(() => "{}")).name;
      if (!themeName) {
        throw Error(`${manifestPath} missing or malformed.`);
      }
      if (!await fileExists(cssPath)) {
        throw Error(`${cssPath} missing.`);
      }
      const themeDest = _path2.default.join(obsidianDir, "themes", themeName);
      await _promises2.default.mkdir(themeDest, { recursive: true });
      await linkOrCp(manifestPath, _path2.default.join(themeDest, "manifest.json"));
      await linkOrCp(cssPath, _path2.default.join(themeDest, "theme.css"));
      if (enabledTheme && enabled) {
        throw Error("You can only have one enabled theme.");
      } else if (enabled) {
        enabledTheme = themeName;
      }
    }
    if (themes.length > 0) {
      const appearancePath = _path2.default.join(obsidianDir, "appearance.json");
      let appearance = {};
      if (await fileExists(appearancePath)) {
        appearance = JSON.parse(await _promises2.default.readFile(appearancePath, "utf-8"));
      }
      appearance.cssTheme = _nullishCoalesce(enabledTheme, () => ( ""));
      await _promises2.default.writeFile(appearancePath, JSON.stringify(appearance, void 0, 2));
    }
  }
  /**
   * Sets up the config dir to use for the `--user-data-dir` in obsidian. Returns the path to the created config dir.
   *
   * @param params.appVersion Obsidian app version
   * @param params.installerVersion Obsidian version string.
   * @param params.appPath Path to the asar file to install. Will download if omitted.
   * @param params.vault Path to the vault to open in Obsidian
   * @param params.localStorage items to add to localStorage. `$vaultId` in the keys will be replaced with the vaultId
   * @param params.chromePreferences Chrome preferences to add to the Preferences file
   */
  async setupConfigDir(params) {
    const [appVersion, installerVersion] = await this.resolveVersion(params.appVersion, params.installerVersion);
    const configDir = await makeTmpDir("obsidian-launcher-config-");
    const vaultId = _crypto2.default.randomBytes(8).toString("hex");
    const localStorageData = {
      "most-recently-installed-version": appVersion,
      // prevents the changelog page on boot
      [`enable-plugin-${vaultId}`]: "true",
      // Disable "safe mode" and enable plugins
      ..._lodash2.default.mapKeys(_nullishCoalesce(params.localStorage, () => ( {})), (v, k) => k.replace("$vaultId", _nullishCoalesce(vaultId, () => ( ""))))
    };
    const chromePreferences = _lodash2.default.merge(
      // disables the "allow pasting" bit in the dev tools console
      { "electron": { "devtools": { "preferences": {
        // chrome switched between using kebab-case and CamelCase sometime between 114.0.5735.289 and 120.0.6099.283
        "disable-self-xss-warning": "true",
        "disableSelfXssWarning": "true"
      } } } },
      _nullishCoalesce(params.chromePreferences, () => ( {}))
    );
    const obsidianJson = {
      updateDisabled: true,
      // prevents Obsidian trying to auto-update on boot.
      cli: true
      // enable the CLI
    };
    if (params.vault !== void 0) {
      if (!await fileExists(params.vault)) {
        throw Error(`Vault path ${params.vault} doesn't exist.`);
      }
      Object.assign(obsidianJson, {
        vaults: {
          [vaultId]: {
            path: _path2.default.resolve(params.vault),
            ts: (/* @__PURE__ */ new Date()).getTime(),
            open: true
          }
        }
      });
      if (_semver2.default.lt(appVersion, "0.10.0")) {
        obsidianJson.last_open = vaultId;
      }
    }
    await _promises2.default.writeFile(_path2.default.join(configDir, "obsidian.json"), JSON.stringify(obsidianJson));
    await _promises2.default.writeFile(_path2.default.join(configDir, "Preferences"), JSON.stringify(chromePreferences));
    let appPath = params.appPath;
    if (!appPath) {
      appPath = await this.downloadApp(appVersion);
    }
    await linkOrCp(appPath, _path2.default.join(configDir, _path2.default.basename(appPath)));
    const localStorage = new ChromeLocalStorage(configDir);
    await localStorage.setItems("app://obsidian.md", localStorageData);
    await localStorage.close();
    return configDir;
  }
  /**
   * Sets up a vault for Obsidian, installing plugins and themes and optionally copying the vault to a temporary
   * directory first.
   * @param params.vault Path to the vault to open in Obsidian
   * @param params.copy Whether to copy the vault to a tmpdir first. Default false
   * @param params.plugins List of plugins to install in the vault
   * @param params.themes List of themes to install in the vault
   * @returns Path to the copied vault (or just the path to the vault if copy is false)
   */
  async setupVault(params) {
    let vault = params.vault;
    if (params.copy) {
      const dest = await makeTmpDir(`${_path2.default.basename(vault)}-`);
      await _promises2.default.cp(vault, dest, { recursive: true, preserveTimestamps: true });
      vault = dest;
    }
    await this.installPlugins(vault, _nullishCoalesce(params.plugins, () => ( [])));
    await this.installThemes(vault, _nullishCoalesce(params.themes, () => ( [])));
    return vault;
  }
  /**
   * Downloads and launches Obsidian with a sandboxed config dir and a specifc vault open. Optionally install plugins
   * and themes first.
   * 
   * @param params.appVersion Obsidian app version. Default "latest"
   * @param params.installerVersion Obsidian installer version. Default "latest"
   * @param params.vault Path to the vault to open in Obsidian
   * @param params.copy Whether to copy the vault to a tmpdir first. Default false
   * @param params.plugins List of plugins to install in the vault
   * @param params.themes List of themes to install in the vault
   * @param params.args CLI args to pass to Obsidian
   * @param params.localStorage items to add to localStorage. `$vaultId` in the keys will be replaced with the vaultId
   * @param params.spawnOptions Options to pass to `spawn`
   * @returns The launched child process and the created tmpdirs
   */
  async launch(params) {
    const [appVersion, installerVersion] = await this.resolveVersion(
      _nullishCoalesce(params.appVersion, () => ( "latest")),
      _nullishCoalesce(params.installerVersion, () => ( "latest"))
    );
    const appPath = await this.downloadApp(appVersion);
    const installerPath = await this.downloadInstaller(installerVersion);
    let vault = params.vault;
    if (vault) {
      vault = await this.setupVault({
        vault,
        copy: _nullishCoalesce(params.copy, () => ( false)),
        plugins: params.plugins,
        themes: params.themes
      });
    }
    const configDir = await this.setupConfigDir({
      appVersion,
      installerVersion,
      appPath,
      vault,
      localStorage: params.localStorage
    });
    const proc = _child_process2.default.spawn(installerPath, [
      `--user-data-dir=${configDir}`,
      // Workaround for SUID issue on linux. See https://github.com/electron/electron/issues/42510
      ...process.platform == "linux" ? ["--no-sandbox"] : [],
      ..._nullishCoalesce(params.args, () => ( [])),
      "--tag=obsidian-launcher"
      // hack so we can identify obsidian-launcher processes when connecting the CLI
    ], {
      ...params.spawnOptions
    });
    return { proc, configDir, vault };
  }
  /** 
   * Updates the info in obsidian-versions.json. The obsidian-versions.json file is used in other launcher commands
   * and in wdio-obsidian-service to get metadata about Obsidian versions in one place such as minInstallerVersion and
   * the internal Electron version.
   */
  async updateVersionList(original, opts = {}) {
    return updateObsidianVersionList(original, { maxInstances: opts.maxInstances });
  }
  /**
   * Returns true if the Obsidian version is already in the cache.
   * @param type on of "app" or "installer"
   * @param version Obsidian app/installer version
   */
  async isInCache(type, version) {
    version = (await this.getVersionInfo(version)).version;
    let dest;
    if (type == "app") {
      dest = `obsidian-app/obsidian-${version}.asar`;
    } else {
      const { platform, arch } = process;
      dest = `obsidian-installer/${platform}-${arch}/Obsidian-${version}`;
    }
    return await fileExists(_path2.default.join(this.cacheDir, dest));
  }
  /**
   * Returns true if we either have the credentials to download the version or it's already in cache.
   * This is only relevant for Obsidian beta versions, as they require Obsidian insider credentials to download.
   * @param appVersion Obsidian app version
   */
  async isAvailable(appVersion) {
    const versionInfo = await this.getVersionInfo(appVersion);
    if (!versionInfo.downloads.asar || !versionInfo.minInstallerVersion) {
      return false;
    }
    if (new URL(versionInfo.downloads.asar).hostname.endsWith(".obsidian.md")) {
      const hasCreds = !!(process.env["OBSIDIAN_EMAIL"] && process.env["OBSIDIAN_PASSWORD"]) || await fileExists(_path2.default.join(this.cacheDir, "obsidian-credentials.env"));
      const inCache = await this.isInCache("app", versionInfo.version);
      return hasCreds || inCache;
    } else {
      return true;
    }
  }
  /**
   * Return the command needed to run the Obsidian CLI.
   * 
   * As obsidian-launcher sandboxes the config dir for each Obsidian instance, the Obsidian CLI won't connect to the
   * launched instances by default. This method takes Obsidian CLI args, and then returns an [executable, args] tuple
   * that can be used to launch the Obsidian CLI against the sandboxed instances.
   * 
   * Like the regular Obsidian CLI, it will connect to the instance matching the `vault=` argument if present, or
   * the cwd.
   * 
   * Just pass the result to child_process.spawn or child_process.execFile to run the command.
   * 
   * Example:
   * ```js
   * import child_process from "child_process";
   * import util from "util";
   * const execFile = util.promisify(child_process.execFile);
   * const [executable, args] = await launcher.getObsidianCli(["file", "file=Dashboard"]);
   * const {stdout, stderr} = await execFile(executable, args);
   * ```
   * 
   * The Obsidian CLI only works on Obsidian >=1.12.0 with installer >=1.11.7.
   * See https://help.obsidian.md/cli
   * 
   * @returns [executable, args] tuple
   */
  async getObsidianCli(args) {
    return await getObsidianCli(args);
  }
};






exports.consola = consola; exports.watchFiles = watchFiles; exports.minSupportedObsidianVersion = minSupportedObsidianVersion; exports.ObsidianLauncher = ObsidianLauncher;
